{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Ejercicio: Alinemientos\n",
    "\n",
    "### Integrantes\n",
    "- Ángel de la Cruz\n",
    "- Otokani Ortega\n",
    "- Paulina Rosales \n",
    "- Dante Torres"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Blast import NCBIWWW\n",
    "\n",
    "fasta_string = open(\"../docs/unknown.fasta\").read()\n",
    "result_handle = NCBIWWW.qblast(\"tblastn\", \"nt\", fasta_string)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sequence: gb|EFC5411094.1| tryptophan permease [Escherichia coli] >gb|EFI1481481.1| tryptophan permease [Escherichia coli] \n",
      "\n",
      "sequence: gb|QLU08459.1| tryptophan permease [Escherichia coli] \n",
      "\n",
      "sequence: ref|WP_097339920.1| tryptophan permease [Escherichia coli] >gb|EGO8886786.1| tryptophan permease [Escherichia coli] >gb|EGO9479947.1| tryptophan permease [Escherichia coli] >gb|EGP6105948.1| tryptophan permease [Escherichia coli] \n",
      "\n",
      "sequence: ref|WP_137544726.1| tryptophan permease [Escherichia coli] >dbj|GDN36112.1| tryptophan-specific transport protein [Escherichia coli] \n",
      "\n",
      "sequence: ref|WP_111961228.1| tryptophan permease [Escherichia coli] >emb|SQY27095.1| tryptophan-specific transport protein [Escherichia coli] >emb|SQY69104.1| tryptophan-specific transport protein [Escherichia coli] >emb|SQZ07686.1| tryptophan-specific transport protein [Escherichia coli] \n",
      "\n"
     ]
    }
   ],
   "source": [
    "from Bio.Blast import NCBIXML\n",
    "\n",
    "\n",
    "result_handle = open(\"../docs/ZVJ05K44013-Alignment.xml\")\n",
    "for blast_result in NCBIXML.parse(result_handle):\n",
    "    for desc in blast_result.descriptions:\n",
    "        if desc.e < 1e-5:\n",
    "            print('sequence:', desc.title, \"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "IGXGMXXXPXXXXGXXF\n"
     ]
    }
   ],
   "source": [
    "from Bio import AlignIO\n",
    "from Bio.Align import AlignInfo\n",
    "\n",
    "al = []\n",
    "alignments = AlignIO.read(\"../docs/ejercicio.phy\", \"phylip\")\n",
    "gap = alignments[0].seq.find('-')\n",
    "alineamiento = alignments[:,0:gap]\n",
    "\n",
    "summary_align = AlignInfo.SummaryInfo(alineamiento)\n",
    "consensus = summary_align.dumb_consensus()\n",
    "sec_consenso = consensus\n",
    "print(sec_consenso)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "from urllib.request import urlopen\n",
    "from urllib.parse import urlencode\n",
    "\n",
    "def scan(seq=\"\", mirror=\"https://prosite.expasy.org\", output=\"xml\", **keywords):\n",
    "    \n",
    "    #Borré la descripción de la función solo para que la visualización de los cambios fuera más sencilla\n",
    "    parameters = {\"seq\": seq, \"output\": output}\n",
    "    \n",
    "    ##Cambio para solo usar sig\n",
    "    if seq is None:\n",
    "        parameters = {\"output\": output}\n",
    "    ##\n",
    "    \n",
    "    for key, value in keywords.items():\n",
    "        if value is not None:\n",
    "            parameters[key] = value\n",
    "\n",
    "        \n",
    "    command = urlencode(parameters)\n",
    "    url = \"%s/cgi-bin/prosite/PSScan.cgi?%s\" % (mirror, command)\n",
    "    print(url)\n",
    "    handle = urlopen(url)\n",
    "    return handle"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "https://prosite.expasy.org/cgi-bin/prosite/PSScan.cgi?output=xml&sig=I-G-X-G-M-x%283%29-P-x%284%29-G-x%282%29-F\n",
      "Number of matches: 11\n",
      "{'sequence_ac': 'P0AAD2', 'sequence_id': 'MTR_ECOLI', 'sequence_db': 'sp', 'start': 25, 'stop': 41, 'signature_ac': 'USERPAT1'}\n",
      "{'sequence_ac': 'P44614', 'sequence_id': 'MTR_HAEIN', 'sequence_db': 'sp', 'start': 21, 'stop': 37, 'signature_ac': 'USERPAT1'}\n",
      "{'sequence_ac': 'P0AAD3', 'sequence_id': 'MTR_SHIFL', 'sequence_db': 'sp', 'start': 25, 'stop': 41, 'signature_ac': 'USERPAT1'}\n"
     ]
    }
   ],
   "source": [
    "from Bio.ExPASy import ScanProsite\n",
    "\n",
    "\n",
    "PROSITE_consensus = \"I-G-X-G-M-x(3)-P-x(4)-G-x(2)-F\" \n",
    "handle = scan(seq = None , sig=PROSITE_consensus)\n",
    "results = ScanProsite.read(handle)\n",
    "\n",
    "print (\"Number of matches:\", results.n_match)\n",
    "n = 0\n",
    "for r in results:\n",
    "    print(r)\n",
    "    n += 1\n",
    "    if n > 2:\n",
    "        break\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "https://prosite.expasy.org/cgi-bin/prosite/PSScan.cgi?output=xml&seq=P0AAD2\n",
      "{'signature_id': 'AROMATIC_AA_PERMEASE_1', 'sequence_db': 'sp', 'level_tag': '(0)', 'sequence_id': 'MTR_ECOLI', 'signature_ac': 'PS00594', 'start': 25, 'sequence_ac': 'P0AAD2', 'stop': 41}\n",
      "\n",
      " AROMATIC_AA_PERMEASE_1\n",
      "I-G-[GA]-G-M-[LF]-[SA]-x-P-x(3)-[SA]-G-x(2)-F.\n"
     ]
    }
   ],
   "source": [
    "from Bio.ExPASy import ScanProsite\n",
    "from Bio import ExPASy\n",
    "from Bio.ExPASy import Prosite\n",
    "\n",
    "\n",
    "uniprot_id = \"P0AAD2\" \n",
    "handle = scan(seq=uniprot_id)\n",
    "results = ScanProsite.read(handle)\n",
    "for result in results:\n",
    "    print(result)\n",
    "    acc_numba = result['signature_ac']\n",
    "    \n",
    "handle = ExPASy.get_prosite_raw(acc_numba)\n",
    "record = Prosite.read(handle)\n",
    "print (\"\\n\", record.name)\n",
    "print (record.pattern)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'PDOC00513'"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "record.pdoc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "********************************************\n",
      "* Aromatic amino acids permeases signature *\n",
      "********************************************\n",
      "\n",
      "It has been shown [1] that some proteins involved in the transport of aromatic\n",
      "amino acids in Escherichia coli and related bacteria are evolutionary related;\n",
      "these permeases are:\n",
      "\n",
      " - Tryptophan-specific transport protein (tryptophan permease) (gene mtr).\n",
      " - Low affinity tryptophan permease (gene tnaB or trpP).\n",
      " - Tyrosine-specific transport protein (tyrosine permease) (gene tyrP).\n",
      " - Tyrosine permease from Erwinia herbicola (gene: tutB).\n",
      "\n",
      "These permeases are proteins of  about 400 to 420 amino  acids  which probably\n",
      "contain 11 or 12 transmembrane regions. The best conserved domain is a stretch\n",
      "of 20 residues which seems to be  located  in a cytoplasmic  loop  between the\n",
      "first and second transmembrane region.\n",
      "\n",
      "-Consensus pattern: I-G-[GA]-G-M-[LF]-[SA]-x-P-x(3)-[SA]-G-x(2)-F\n",
      "-Sequences known to belong to this class detected by the pattern: ALL.\n",
      "-Other sequence(s) detected in Swiss-Prot: NONE.\n",
      "-Last update: November 1997 / Pattern and text revised.\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from Bio.ExPASy import Prodoc\n",
    "\n",
    "handle = ExPASy.get_prosite_raw(\"PDOC00513\")\n",
    "record = Prodoc.read(handle)\n",
    "\n",
    "record.text\n",
    "print(record.text.replace('\\\\n','\\n'))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
