{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Tarea 3: *Bases de datos*\n",
    "\n",
    "***Paulina Rosales Becerra***\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='indice'></a>\n",
    "## Índice\n",
    "1. [Ejercicio 1](#1)\n",
    "1. [Ejercicio 2](#2)\n",
    "1. [Ejercicio 3](#3)\n",
    "1. [Ejercicio 4](#4)\n",
    "\n",
    "**Nota:** La mayoría de los códigos cuentan con variables *filename* y *structure_id* para poder editar la ruta del archivo *input* o *output* en caso de ser requerido."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='1'></a>\n",
    "## Ejercicio 1\n",
    "[Regresar al índice](#indice)\n",
    "\n",
    "<p>Arreglar el código para que solo muestre las cominaciones únicas</p>\n",
    "\n",
    "\n",
    "<h3 id=\"sample-output\">Resultado emitido de muestra</h3>\n",
    "<div class=\"codehilite\"><pre>possible di-sulfide bond: Cys 155 - Cys 181 2.0\n",
    "possible di-sulfide bond: Cys 257 - Cys 338 2.03\n",
    "possible di-sulfide bond: Cys 515 - Cys 541 2.02\n",
    "possible di-sulfide bond: Cys 618 - Cys 699 2.03\n",
    "possible di-sulfide bond: Cys 881 - Cys 855 2.02\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "possible di-sulfide bond: Cys 155 - Cys 155 0.0\n",
      "possible di-sulfide bond: Cys 155 - Cys 181 2.0\n",
      "possible di-sulfide bond: Cys 181 - Cys 155 2.0\n",
      "possible di-sulfide bond: Cys 181 - Cys 181 0.0\n",
      "possible di-sulfide bond: Cys 221 - Cys 221 0.0\n",
      "possible di-sulfide bond: Cys 257 - Cys 257 0.0\n",
      "possible di-sulfide bond: Cys 257 - Cys 338 2.03\n",
      "possible di-sulfide bond: Cys 319 - Cys 319 0.0\n",
      "possible di-sulfide bond: Cys 338 - Cys 257 2.03\n",
      "possible di-sulfide bond: Cys 338 - Cys 338 0.0\n",
      "possible di-sulfide bond: Cys 515 - Cys 515 0.0\n",
      "possible di-sulfide bond: Cys 515 - Cys 541 2.02\n",
      "possible di-sulfide bond: Cys 541 - Cys 515 2.02\n",
      "possible di-sulfide bond: Cys 541 - Cys 541 0.0\n",
      "possible di-sulfide bond: Cys 618 - Cys 618 0.0\n",
      "possible di-sulfide bond: Cys 618 - Cys 699 2.03\n",
      "possible di-sulfide bond: Cys 680 - Cys 680 0.0\n",
      "possible di-sulfide bond: Cys 699 - Cys 618 2.03\n",
      "possible di-sulfide bond: Cys 699 - Cys 699 0.0\n",
      "possible di-sulfide bond: Cys 855 - Cys 855 0.0\n",
      "possible di-sulfide bond: Cys 855 - Cys 881 2.02\n",
      "possible di-sulfide bond: Cys 881 - Cys 855 2.02\n",
      "possible di-sulfide bond: Cys 881 - Cys 881 0.0\n",
      "possible di-sulfide bond: Cys 1021 - Cys 1021 0.0\n"
     ]
    }
   ],
   "source": [
    "# CODIGO ORIGINAL\n",
    "import Bio.PDB.PDBParser\n",
    "\n",
    "# Use QUIET=True to avoid lots of warnings...\n",
    "# Ruta al archivo 1kcw.pdb\n",
    "parser = Bio.PDB.PDBParser(QUIET=True)\n",
    "structure = parser.get_structure(\"1KCW\", \"../source/1kcw.pdb\")\n",
    "achain = structure[0]['A']\n",
    "\n",
    "cysresidues = []\n",
    "\n",
    "#Voy por cada residuo hasta encontrar CYS\n",
    "for residue in achain:\n",
    "    if residue.get_resname() == 'CYS':\n",
    "        cysresidues.append(residue)\n",
    "        \n",
    "# Veo todos los residuos y sus combinaciones. Como vimos anteriormente, al restar átomos nos da la distancia\n",
    "for c1 in cysresidues:\n",
    "    c1index = c1.get_id()[1]\n",
    "    for c2 in cysresidues:\n",
    "        c2index = c2.get_id()[1]\n",
    "        #Estoy enteresada en el átomo de SG \"sulfur-gamma\"\n",
    "        if (c1['SG'] - c2['SG']) < 8.0:\n",
    "            print(\"possible di-sulfide bond:\", end=\" \")\n",
    "            print(\"Cys\",c1index,\"-\", end=\" \")\n",
    "            print(\"Cys\",c2index, end=\" \")\n",
    "            print(round(c1['SG'] - c2['SG'],2))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Possible di-sulfide bond: Cys 155 - Cys 181 2.0\n",
      "Possible di-sulfide bond: Cys 257 - Cys 338 2.03\n",
      "Possible di-sulfide bond: Cys 515 - Cys 541 2.02\n",
      "Possible di-sulfide bond: Cys 618 - Cys 699 2.03\n",
      "Possible di-sulfide bond: Cys 855 - Cys 881 2.02\n"
     ]
    }
   ],
   "source": [
    "import Bio.PDB.PDBParser\n",
    "\n",
    "filename = \"../source/1kcw.pdb\"\n",
    "structure_id = \"1kcw\"\n",
    "parser = Bio.PDB.PDBParser(QUIET = True)\n",
    "structure = parser.get_structure(structure_id, filename)\n",
    "chain_A = structure[0]['A']\n",
    "\n",
    "cysresidues = []\n",
    "\n",
    "# Extract all cystein residues\n",
    "for residue in chain_A:\n",
    "    if residue.get_resname() == 'CYS':\n",
    "        cysresidues.append(residue)\n",
    "        \n",
    "# Generate all cystein residues comparition\n",
    "for cys_1 in cysresidues:\n",
    "    cys1_index = cys_1.get_id()[1]\n",
    "    for cys_2 in cysresidues:\n",
    "        cys2_index = cys_2.get_id()[1]\n",
    "        # Focus on 'Sulfur gamma' atom (SG) to consider distances between linked cysteins\n",
    "        if (cys_1['SG'] - cys_2['SG']) < 8.0 and cys1_index < cys2_index:\n",
    "            print(\"Possible di-sulfide bond:\", end=\" \")\n",
    "            print(\"Cys\",cys1_index,\"-\", end=\" \")\n",
    "            print(\"Cys\",cys2_index, end=\" \")\n",
    "            print(round(cys_1['SG'] - cys_2['SG'],2))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='2'></a>\n",
    "## Ejercicio 2\n",
    "[Regresar al índice](#indice)\n",
    "\n",
    "<p> BS3 es un  agente entrecruzante  impermeable a la membrana plasmática. Sirve para el estudio de\n",
    "interacciones de proteínas de membrana plasmática con residuos de lisina\n",
    "accesibles desde el medio extracelular</p>\n",
    "+ Encuentren los pares de lisinas que podrían ser unidas utilizando BS3 \n",
    "\n",
    "La distancia suele ser menor a 12 Å"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Possible bond using BS3  Lys 12 - Lys 7 11.29\n"
     ]
    }
   ],
   "source": [
    "#import Bio.PDB.PDBParser\n",
    "\n",
    "filename = \"../source/1ce0.pdb\"\n",
    "structure_id = \"1ce0\"\n",
    "parser = Bio.PDB.PDBParser(QUIET = True)\n",
    "structure = parser.get_structure(structure_id, filename)\n",
    "chain_A = structure[0]['A']\n",
    "chain_B = structure[0]['B']\n",
    "\n",
    "lysresidues_A = []\n",
    "lysresidues_B = []\n",
    "\n",
    "# Generate list containing al lysines fron A chain\n",
    "for residue in chain_A:\n",
    "    if residue.get_resname() == 'LYS':\n",
    "        lysresidues_A.append(residue)\n",
    "        \n",
    "# Generate list containing al lysines fron B chain\n",
    "for residue in chain_B:\n",
    "    if residue.get_resname() == 'LYS':\n",
    "        lysresidues_B.append(residue)\n",
    "        \n",
    "for lys_A in lysresidues_A:\n",
    "    lys_A_NZ = lys_A['NZ']\n",
    "    lys_A_NZ_index = lys_A.get_id()[1]\n",
    "    for lys_B in lysresidues_B:\n",
    "        lys_B_NZ = lys_B['NZ']\n",
    "        lys_B_NZ_index = lys_B.get_id()[1]\n",
    "        #Estoy enteresada en el átomo de SG \"sulfur-gamma\"\n",
    "        if (lys_A_NZ - lys_B_NZ) < 12.0:\n",
    "            print(\"Possible bond using BS3 \", end = \" \")\n",
    "            print(\"Lys\",lys_A_NZ_index,\"-\", end = \" \")\n",
    "            print(\"Lys\",lys_B_NZ_index, end = \" \")\n",
    "            print(round(lys_A_NZ - lys_B_NZ,2))\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='3'></a>\n",
    "## Ejercicio 3\n",
    "[Regresar al índice](#indice)\n",
    "\n",
    "Utilizando Entrez.einfo y Entrez.read, imprime la descripción de:\n",
    "    \n",
    "    \n",
    "- field **'ECNO'**\n",
    "\n",
    "- link **'protein_protein_small_genome'**\n",
    "    \n",
    "    \n",
    "de la base de datos 'protein'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ECNO's description:\n",
      "   EC number for enzyme or CAS registry number\n",
      "\n",
      "protein_protein_small_genome's description: \n",
      "   All proteins from this genome\n"
     ]
    }
   ],
   "source": [
    "from Bio import Entrez\n",
    "\n",
    "Entrez.email = \"prosales@lcg.unam.mx\" \n",
    "handle = Entrez.einfo(db = 'protein') \n",
    "record = Entrez.read(handle) \n",
    "\n",
    "# Move through all fields\n",
    "for field in record[\"DbInfo\"][\"FieldList\"]:\n",
    "    # Get ECNO info\n",
    "    if field['Name'] == 'ECNO':\n",
    "        print(\"%(Name)s's description:\\n   %(Description)s\" % field)\n",
    "\n",
    "# Move through all links\n",
    "for link in record[\"DbInfo\"][\"LinkList\"]:\n",
    "    # Get protein_protein_samll_genome info\n",
    "    if link['Name'] == 'protein_protein_small_genome':\n",
    "        print(\"\\n%(Name)s's description: \\n   %(Description)s\" % link)\n",
    "        \n",
    "handle.close() "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='4'></a>\n",
    "## Ejercicio 4\n",
    "[Regresar al índice](#indice)\n",
    "\n",
    "<p>Automaticen para buscar los siguientes genes en la base de datos nucleotide y guarden las secuencias de las proteínas en un archivo fasta: \n",
    "</p>\n",
    "\n",
    "Organismo | Genes | Aislado\n",
    ":-------- | :---: | :-----:\n",
    "Notoryctes typhlops |PNOC y RUNX2 |-\n",
    "Cypripedium guttatum  |matK |-\n",
    "Hepatitis B virus |X |migrant 467 \n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "USAGE: Input the outpu fasta file name. Indicate the quantity of items to search and the next fields when indicated:\n",
      "\tOrganism: only 1\n",
      "\tGene(s): separated by a space\n",
      "\tIsolate: if just push Enter\n",
      "Enter the output's file name:  ../tmp/Tarea3Ej4.fasta\n",
      "\n",
      "How many items to consider search:  3\n",
      "\n",
      "Organism: Notoryctes typhlops\n",
      "Gene(s): PNOC RUNX2\n",
      "Isolate:  \n",
      "\n",
      "Organism: Cypripedium guttatum\n",
      "Gene(s): matK\n",
      "Isolate:  \n",
      "\n",
      "Organism: Hepatitis B virus\n",
      "Gene(s): X\n",
      "Isolate:  migrant 467\n",
      "\n",
      "Downloading data from (Notoryctes typhlops[Organism]) AND (PNOC[Gene Name] OR RUNX2[Gene Name])\n",
      "Going to download record 1 to 2\n",
      "\n",
      "Downloading data from (Cypripedium guttatum[Organism]) AND matK[Gene Name]\n",
      "Going to download record 1 to 1\n",
      "\n",
      "Downloading data from (Hepatitis B virus[Organism]) AND X[Gene Name] AND migrant 467[Isolate]\n",
      "Going to download record 1 to 1\n",
      "\n",
      "Download finished, fetched sequences saved at ../tmp/Tarea3Ej4.fasta\n"
     ]
    }
   ],
   "source": [
    "from Bio import Entrez\n",
    "Entrez.email = \"prosales@lcg.unam.mx\"\n",
    "from urllib.error import HTTPError \n",
    "\n",
    "print(\"USAGE: Input the outpu fasta file name. Indicate the quantity of items to search and the next fields when indicated:\\n\\tOrganism: only 1\\n\\tGene(s): separated by a space\\n\\tIsolate: if just push Enter\")\n",
    "batch_size = 3\n",
    "\n",
    "out_file = input(\"Enter the output's file name:  \")\n",
    "out_handle = open(out_file, \"w\")\n",
    "\n",
    "# Initialice list containing all needed queries\n",
    "number_search = int(input(\"\\nHow many items to consider search:  \"))\n",
    "all_searchs = []\n",
    "\n",
    "# Generate list containing correct sintax for each query from inputed data\n",
    "for i in range(0, number_search):\n",
    "    search = \"\"\n",
    "    flag = 1\n",
    "    # Data inputs\n",
    "    organism = input(\"\\nOrganism: \")\n",
    "    genes = input(\"Gene(s): \").split()\n",
    "    isolate = input(\"Isolate:  \")\n",
    "    search = \"(\" + organism + \"[Organism])\"\n",
    "    # Split genes with the rigth sintax in case of multiple inputed genes\n",
    "    if len(genes) == 1:\n",
    "        search += \" AND \" + genes[0] + \"[Gene Name]\"\n",
    "    elif len(genes) > 1:\n",
    "        search += \" AND (\"\n",
    "        for gene in genes:\n",
    "            if flag == 1:\n",
    "                search += gene + \"[Gene Name]\"\n",
    "                flag += 1\n",
    "            elif flag > 1 and flag != len(genes):\n",
    "                search += \" OR \" + gene + \"[Gene Name]\"\n",
    "                flag += 1\n",
    "            elif flag == len(genes):\n",
    "                search += \" OR \" + gene + \"[Gene Name])\"\n",
    "            \n",
    "    # Add isolate field if inputed\n",
    "    if isolate != \"\":\n",
    "        search += \" AND \" + isolate + \"[Isolate]\"\n",
    "    # Add it to the list\n",
    "    all_searchs.append(search)\n",
    "            \n",
    "# Search queries one by one in Nucleotide database\n",
    "for search in all_searchs:\n",
    "    print('\\nDownloading data from {}'.format(search))\n",
    "    search_handle = Entrez.esearch(db = \"nucleotide\", term = search, usehistory = 'y')\n",
    "    search_results = Entrez.read(search_handle)\n",
    "    search_handle.close()\n",
    "    count = int(search_results[\"Count\"])\n",
    "    webenv = search_results[\"WebEnv\"]\n",
    "    query_key = search_results[\"QueryKey\"]\n",
    "    # Download particular records\n",
    "    for start in range(0, count, batch_size):\n",
    "        end = min(count, start+batch_size)\n",
    "        print(\"Going to download record %i to %i\" % (start+1, end))\n",
    "        attempt = 1\n",
    "        while attempt <= 3:\n",
    "            try:\n",
    "                fetch_handle = Entrez.efetch(db=\"nucleotide\", rettype=\"fasta\", retmode=\"text\",\n",
    "                                             retstart=start, retmax=batch_size,\n",
    "                                             webenv=webenv, query_key=query_key)\n",
    "\n",
    "                break\n",
    "            # Error exception\n",
    "            except HTTPError as err:\n",
    "                if 500 <= err.code <= 599:\n",
    "                    print(\"Received error from server %s\" % err)\n",
    "                    print(\"Attempt %i of 3\" % attempt)\n",
    "                    attempt += 1\n",
    "                    time.sleep(15)\n",
    "                else:\n",
    "                    raise\n",
    "                \n",
    "    data = fetch_handle.read()\n",
    "    fetch_handle.close()\n",
    "    out_handle.write(data)\n",
    "    \n",
    "print('\\nDownload finished, fetched sequences saved at {}'.format(out_file))\n",
    "out_handle.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">KY938699.1 Notoryctes typhlops runt-related transcription factor 2 (RUNX2) gene, partial cds\n",
      "ACCCGGCAGATGAGCGACGTGAGCCCGGTAGTGGCTGCCCAGCAGCAGCAGCAGCAGCAGCAGCAGCAAC\n",
      "AGCAACAGCAGCAACAGCAGCAGCAGCAGCAGCAGCAGCAGGAGGCGGCGGCCGCAGCAGCGGCGGCGGC\n",
      "GGCGGCGGCGGCGGCGGCGGCAGCGGCAGCGGCAGCCGCGGTGCCGAGGCTGCGCCCGCCTCACGACAAC\n",
      "CGCACGATG\n",
      "\n",
      ">JN633765.1 Notoryctes typhlops prepronociceptin (PNOC) gene, partial cds\n",
      "TTCTGGGCTCCTTGCATCAAGGCCATGAATAAGGGGCCCTGGCAGCTCAGCTCAGCCAGCAAGGACAAGG\n",
      "CAATGACATCCTTCTATCAGCCACAGTTGATGGAGGCACAATCCCTAAGCCAGATGGAGGGCAAGAACCG\n",
      "AGTGTCTCGTGTCAGAAGTTTGTTCCATGCCCAGGAACAAGAGGCAGCAGAAGGCTTAGGGGAGTCAGGG\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "head ../tmp/Tarea3Ej4.fasta"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
