{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Tarea 2: *Comandos Biopython*\n",
    "\n",
    "***Paulina Rosales Becerra***\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='indice'></a>\n",
    "## Índice\n",
    "1. [Ejercicio 1](#1)\n",
    "1. [Ejercicio 2](#2)\n",
    "1. [Ejercicio 3](#3)\n",
    "1. [Ejercicio 4](#4)\n",
    "1. [Ejercicio 5](#5)\n",
    "1. [Ejercicio 6](#6)\n",
    "\n",
    "**Nota:** La mayoría de los códigos cuentan con variables *filename* para poder editar la ruta del archivo *input* en caso de ser requerido."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='1'></a>\n",
    "## Ejercicio 1\n",
    "[Regresar al índice](#indice)\n",
    "\n",
    "Un ORF inicia con un codón inicial y termina, ya sea con un codón final o al final de la cádena.\n",
    "Utilizando como input una cadena de ADN, obtener la cadena proteica de mayor longitud que puede ser traducida de cualquiera de sus ORF.\n",
    "Si más de una cadena protéica de longitud maxima existe, entonces podrás emitir cualquier solución."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {
    "scrolled": true,
    "tags": [
     "raises-exception"
    ]
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This code estimates the largets protein considering all ORFs of a given DNA sequence\n",
      "\n",
      "Enter the DNA sequence:\n",
      "AGCCATGTAGCTAACTCAGGTTACATGGGGATGACCCCGCGACTTGGATTAGAGTCTCTTTTGGAATAAGCCTGAATGATCCGAGTAGCATCTCAG\n",
      "\n",
      "Largest protein considering all ORFs: \n",
      " \tMLLGSFRLIPKETLIQVAGSSPCNLS\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/lib/python3.5/dist-packages/Bio/Seq.py:2859: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.\n",
      "  BiopythonWarning,\n"
     ]
    }
   ],
   "source": [
    "from Bio.Seq import Seq\n",
    "from Bio import SeqIO\n",
    "\n",
    "def obtain_prot(rna):\n",
    "    # Generate all posible (considering 3 ORFs)proteins given a singls RNA strand\n",
    "    prot_vect = []\n",
    "    prot_size_vect = []\n",
    "    nt = 0\n",
    "    flag = True\n",
    "    # Move through the string by single characters\n",
    "    while flag:\n",
    "        # Begin the sequence avoiding last start codon considered\n",
    "        rna = rna[nt:]\n",
    "        start_codon_index = rna.find(\"AUG\")\n",
    "        # If there isn't any start codons left or the sequence is over break the loop\n",
    "        if start_codon_index == -1 or len(rna) < 3:\n",
    "            flag = False\n",
    "        else:\n",
    "            # Generate the protein according to the start codon till the fist stop codon found\n",
    "            rna = rna[start_codon_index:]\n",
    "            prot = rna.translate(to_stop = True)\n",
    "            prot_vect.append(str(prot))\n",
    "            prot_size_vect.append(len(prot))\n",
    "            nt += 3\n",
    "    return (prot_vect, prot_size_vect) \n",
    "       \n",
    "print('This code estimates the largets protein considering all ORFs of a given DNA sequence\\n\\nEnter the DNA sequence:')\n",
    "dna = Seq(input().upper())\n",
    "# DNA to RNA\n",
    "fwd_rna = dna.transcribe()\n",
    "rvrs_rna = fwd_rna.reverse_complement()\n",
    "\n",
    "# Calculate 3 ORF for each RNA strand translation\n",
    "fwd_prot, fwd_prot_size = obtain_prot(fwd_rna)\n",
    "rvrs_prot, rvrs_prot_size = obtain_prot(rvrs_rna)\n",
    "\n",
    "# Get all protein data together\n",
    "all_prot = fwd_prot + rvrs_prot\n",
    "all_prot_sizes = fwd_prot_size + rvrs_prot_size\n",
    "largest_prot = all_prot[all_prot_sizes.index(max(all_prot_sizes))]\n",
    "print('\\nLargest protein considering all ORFs: \\n \\t{}\\n'.format(largest_prot))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='2'></a>\n",
    "## Ejercicio 2\n",
    "[Regresar al índice](#indice)\n",
    "\n",
    "1. Teniendo como input un archivo fasta, crea un diccionario de todas las secuencias en tu archivo FASTA.\n",
    "\n",
    "2. Encuentra todos los codones en tu primer marco de lectura. \n",
    "\n",
    "3. Imprime los codones de cada secuncia en formato FASTA. \n",
    "\n",
    "   ```\n",
    "   >my_seq_marco_1\n",
    "   ATG TTC ATC\n",
    "   ```\n",
    "\n",
    "\n",
    "4. Imprime los codones de cada secuencia en los 6 marcos lectura."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {
    "tags": [
     "raises-exception"
    ]
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "> seq1\n",
      "AAG AGC AGC TCG CGC TAA TGT GAT AGA TGG CGG TAA AGT AAA TGT CCT ATG GGC CAC CAA TTA TGG TGT ATG AGT GAA TCT CTG GTC CGA GAT TCA CTG AGT AAC TGC TGT ACA CAG TAG TAA CAC GTG GAG ATC CCA TAA GCT TCA CGT GTG GTC CAA TAA AAC ACT CCG TTG GTC AAC \n",
      "\n",
      "> seq2\n",
      "GCC ACA GAG CCT AGG ACC CCA ACC TAA CCT AAC CTA ACC TAA CCT ACA GTT TGA TCT TAA CCA TGA GGC TGA GAA GCG ATG TCC TGA CCG GCC TGT CCT AAC CGC CCT GAC CTA ACC GGC TTG ACC TAA CCG CCC TGA CCT AAC CAG GCT AAC CTA ACC AAA CCG TGA AAA AAG GAA TCT \n",
      "\n",
      "> seq3\n",
      "ATG AAA GTT ACA TAA AGA CTA TTC GAT GCA TAA ATA GTT CAG TTT TGA AAA CTT ACA TTT TGT TAA AGT CAG GTA CTT GTG TAT AAT ATC AAC TAA AT \n",
      "\n",
      "> seq4\n",
      "ATG CTA ACC AAA GTT TCA GTT CGG ACG TGT CGA TGA GCG ACG CTC AAA AAG GAA ACA ACA TGC CAA ATA GAA ACG ATC AAT TCG GCG ATG GAA ATC AGA ACA ACG ATC AGT TTG GAA ATC AAA ATA GAA ATA ACG GGA ACG ATC AGT TTA ATA ACA TGA TGC AGA ATA AAG GGA ATA ATC AAT TTA ATC CAG GTA ATC AGA ACA GAG GT \n"
     ]
    }
   ],
   "source": [
    "# from Bio import SeqIO\n",
    "\n",
    "def get_triplets(seq):\n",
    "    # Generate a string with whole fasta sequence splited in triplets\n",
    "    seq_str = str(seq)\n",
    "    divided_seq_vect = [(seq_str[i:i+3]) for i in range(0, len(seq_str), 3)]\n",
    "    divided_seq = ''\n",
    "    for i in divided_seq_vect:\n",
    "        divided_seq += i + ' '\n",
    "    return (divided_seq)\n",
    "    \n",
    "filename = ('../source/seq.nt.fa')\n",
    "for seq_record in SeqIO.parse(filename, \"fasta\"):\n",
    "    print('\\n>',seq_record.id)\n",
    "    divided_seq = get_triplets(seq_record.seq)\n",
    "    print(divided_seq)\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='3'></a>\n",
    "## Ejercicio 3\n",
    "[Regresar al índice](#indice)\n",
    "\n",
    "Obtén el *número* de lecturas cuyo promedio de calidad esté por debajo del umbral dado.\n",
    "\n",
    "*Input:* Un umbral de calidad, junto con un fastq que los contenga.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fastqc file path:\n",
      "../source/sample.fastq\n",
      "\n",
      "Quality threshold to consider:\n",
      "28\n",
      "\n",
      "Total of 30 reads\n",
      "From which 4 reads with mean quality lower than 28 were saved in ../tmp/low_quality.fastq\n"
     ]
    }
   ],
   "source": [
    "# from Bio impor SeqIO\n",
    "\n",
    "def quality_mean(qualities_vect):\n",
    "    partial_sum = 0\n",
    "    for quality in qualities_vect:\n",
    "        partial_sum += quality\n",
    "    mean = partial_sum / len(qualities_vect)\n",
    "    return(mean)\n",
    "  \n",
    "print('Fastq file path:')\n",
    "filename = input()     # used ../source/sample.fastq\n",
    "print('\\nQuality threshold to consider:')\n",
    "threshold = int(input())\n",
    "count = 0\n",
    "\n",
    "# Total reads\n",
    "for record in SeqIO.parse(filename, 'fastq'):\n",
    "    count += 1\n",
    "print('\\nTotal of %i reads' % count)\n",
    "\n",
    "# Save records according to the quality evaluation\n",
    "low_reads = (\n",
    "    record\n",
    "    for record in SeqIO.parse(filename, 'fastq')\n",
    "    if quality_mean(record.letter_annotations['phred_quality']) < threshold\n",
    ")\n",
    "# Count and save low quality records on an output fastq file (extra)\n",
    "count = SeqIO.write(low_reads, '../tmp/low_quality.fastq', 'fastq')\n",
    "print('From which {} reads with mean quality lower than {} were saved in ../tmp/low_quality.fastq'.format(count, threshold))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='4'></a>\n",
    "## Ejercicio 4\n",
    "[Regresar al índice](#indice)\n",
    "\n",
    "El archivo hepatitis.gb contiene información sobre la región de ADN que codifica HBx, una proteína viral de la hepatitis B multifuncional involucrada en la modulación de varias vías al interactuar directa o indirectamente con factores del hospedero (degradación de proteínas, apoptosis, transcripción, transducción de señales, progreso del ciclo celular y estabilidad genética). Escribe un programa de Python que, utilizando este registro de genbank, guarde la secuencia de proteínas correspondiente en formato fasta."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "tags": [
     "raises-exception"
    ]
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saved in ../tmp/hepatitis.fa\n"
     ]
    }
   ],
   "source": [
    "# from Bio import SeqIO\n",
    "\n",
    "filename = '../source/hepatitis.gb'\n",
    "# Save only the record and the sequence\n",
    "prot_seq =(\n",
    "    record\n",
    "    for record in SeqIO.parse(filename, 'genbank')\n",
    "    if record.seq\n",
    ")\n",
    "SeqIO.write(prot_seq, '../tmp/hepatitis.fa', 'fasta')\n",
    "print('Saved as ../tmp/hepatitis.fa')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='5'></a>\n",
    "## Ejercicio 5\n",
    "[Regresar al índice](#indice)\n",
    "\n",
    "Consigue el structure del archivo pdb 4hhb \n",
    "+ Imprime las coordenadas 3D del átomo C-alpha del aminoacido con residuo número 20 en chain C y asigna ese residuo a una variable. \n",
    "+ Calcula el centro geométrico *(centroid)* del aminoácido mencionado anteriormente e imprímelo (esto asume que las masas de los átomos son iguales)\n",
    "    \n",
    "    Para calular el centro geométrico se necesitan sacar las coordenadas de todos los átomos del residuo y sacar el promedio.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Atom 3D coords are: [-26.912   8.767  11.705]\n",
      "Residue centroid =  [-28.167499923706053, 8.430100059509277, 10.92150001525879]\n"
     ]
    }
   ],
   "source": [
    "from Bio import PDB\n",
    "\n",
    "def centroid(coords_sum_vect, residue):\n",
    "    # Calculate residue's centroid assuming same mass for all atoms\n",
    "    mean = []\n",
    "    for coord_sum in coords_sum_vect:\n",
    "        mean.append(coord_sum/len(residue))\n",
    "    return(mean)\n",
    "\n",
    "filename = '../source/4hhb.pdb'\n",
    "structure_id = '4hhb'\n",
    "parser = PDB.PDBParser(QUIET = True)\n",
    "structure = parser.get_structure(structure_id, filename)\n",
    "\n",
    "# Access particular atom\n",
    "wanted_atom = structure[0][\"C\"][20][\"CA\"]\n",
    "residue = structure[0][\"C\"][20]\n",
    "all_atoms_coords = [0, 0, 0]\n",
    "\n",
    "# Sum all atoms' coords \n",
    "for atom in residue:\n",
    "    all_atoms_coords += atom.get_coord() \n",
    "\n",
    "centroid_coords = centroid(all_atoms_coords, residue)\n",
    "\n",
    "print('Atom 3D coords are:', wanted_atom.get_coord())\n",
    "print('Residue centroid = ', centroid_coords)                \n",
    "                "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='6'></a>\n",
    "## Ejercicio 6\n",
    "[Regresar al índice](#indice)\n",
    "\n",
    "1. Usando el archivo test.blastout.xml parsea los resultados e imprime lo siguiente \n",
    "\n",
    "\"queryName\thitID\te-value\thitDesc\tHsp_num\" \n",
    "\n",
    "Si el valor es mejor (menor) que 1e-10. (el query name lo encuentran en el blast_record y el hitDesc dentro del alignment.hit_def)\n",
    "\n",
    "```python\n",
    ">>>blast_record.query_id\n",
    "'Query_1'\n",
    ">>>alignment.hit_id\n",
    "'sp|Q9Y267|S22AE_HUMAN'\n",
    ">>>alignment.hit_def\n",
    "'Solute carrier family 22 member 14 OS=Homo sapiens GN=SLC22A14 PE=2 SV=4'\n",
    ">>>hsp.expect\n",
    "2.35119e-06\n",
    "```\n",
    "\n",
    "\n",
    "|Query Name|Hit ID|e-Value|Hit Description|Hsp Number|\n",
    "|:---------:|:----:|:------:|:----:|:------:|\n",
    "|Query_1|sp\\|Q9Y267\\|S22AE_HUMAN|2.35119e-06|Solute carrier family 22 member 14|1|"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "tags": [
     "raises-exception"
    ]
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Query Name \t Hit ID \t\t e-Value    Hsp Number \t Hit Description \n",
      " ----------------------------------------------------------------------------------------------------------------\n",
      "  Query_1    sp|Q8CFZ5|S22AC_MOUSE    1.47916e-109 \t 1    Solute carrier family 22 member 12 OS=Mus musculus GN=Slc22a12 PE=1 SV=1 \n",
      "\n",
      "  Query_1    sp|Q8CFZ5|S22AC_MOUSE    1.20034e-74 \t 2    Solute carrier family 22 member 12 OS=Mus musculus GN=Slc22a12 PE=1 SV=1 \n",
      "\n",
      "  Query_1    sp|Q3ZAV1|S22AC_RAT    7.69823e-103 \t 1    Solute carrier family 22 member 12 OS=Rattus norvegicus GN=Slc22a12 PE=2 SV=1 \n",
      "\n",
      "  Query_1    sp|Q3ZAV1|S22AC_RAT    1.31588e-62 \t 2    Solute carrier family 22 member 12 OS=Rattus norvegicus GN=Slc22a12 PE=2 SV=1 \n",
      "\n",
      "  Query_1    sp|Q96S37|S22AC_HUMAN    3.00718e-78 \t 1    Solute carrier family 22 member 12 OS=Homo sapiens GN=SLC22A12 PE=1 SV=1 \n",
      "\n",
      "  Query_1    sp|Q96S37|S22AC_HUMAN    4.9569e-46 \t 2    Solute carrier family 22 member 12 OS=Homo sapiens GN=SLC22A12 PE=1 SV=1 \n",
      "\n",
      "  Query_1    sp|Q9NSA0|S22AB_HUMAN    4.379e-59 \t 1    Solute carrier family 22 member 11 OS=Homo sapiens GN=SLC22A11 PE=1 SV=1 \n",
      "\n",
      "  Query_1    sp|Q9NSA0|S22AB_HUMAN    6.63169e-25 \t 2    Solute carrier family 22 member 11 OS=Homo sapiens GN=SLC22A11 PE=1 SV=1 \n",
      "\n",
      "  Query_1    sp|Q2KIV1|S22A9_BOVIN    2.16919e-49 \t 1    Solute carrier family 22 member 9 OS=Bos taurus GN=SLC22A9 PE=2 SV=1 \n",
      "\n",
      "  Query_1    sp|Q2KIV1|S22A9_BOVIN    1.78901e-27 \t 2    Solute carrier family 22 member 9 OS=Bos taurus GN=SLC22A9 PE=2 SV=1 \n",
      "\n",
      "  Query_1    sp|Q8IVM8|S22A9_HUMAN    7.4769e-47 \t 1    Solute carrier family 22 member 9 OS=Homo sapiens GN=SLC22A9 PE=1 SV=1 \n",
      "\n",
      "  Query_1    sp|Q8IVM8|S22A9_HUMAN    7.93154e-18 \t 2    Solute carrier family 22 member 9 OS=Homo sapiens GN=SLC22A9 PE=1 SV=1 \n",
      "\n",
      "  Query_1    sp|Q76M72|S22AR_MOUSE    8.2208e-47 \t 1    Solute carrier family 22 member 27 OS=Mus musculus GN=Slc22a27 PE=1 SV=1 \n",
      "\n",
      "  Query_1    sp|Q76M72|S22AR_MOUSE    4.58388e-20 \t 2    Solute carrier family 22 member 27 OS=Mus musculus GN=Slc22a27 PE=1 SV=1 \n",
      "\n",
      "  Query_1    sp|Q63ZE4|S22AA_HUMAN    9.76472e-47 \t 1    Solute carrier family 22 member 10 OS=Homo sapiens GN=SLC22A10 PE=2 SV=2 \n",
      "\n",
      "  Query_1    sp|Q63ZE4|S22AA_HUMAN    1.21117e-22 \t 2    Solute carrier family 22 member 10 OS=Homo sapiens GN=SLC22A10 PE=2 SV=2 \n",
      "\n",
      "  Query_1    sp|Q6T423|S22AP_HUMAN    1.417e-45 \t 1    Solute carrier family 22 member 25 OS=Homo sapiens GN=SLC22A25 PE=2 SV=2 \n",
      "\n",
      "  Query_1    sp|Q6T423|S22AP_HUMAN    5.2991e-20 \t 2    Solute carrier family 22 member 25 OS=Homo sapiens GN=SLC22A25 PE=2 SV=2 \n",
      "\n",
      "  Query_1    sp|Q8N4F4|S22AO_HUMAN    1.47336e-45 \t 1    Solute carrier family 22 member 24 OS=Homo sapiens GN=SLC22A24 PE=2 SV=1 \n",
      "\n",
      "  Query_1    sp|Q5RC45|S22AO_PONAB    2.77334e-45 \t 1    Solute carrier family 22 member 24 OS=Pongo abelii GN=SLC22A24 PE=2 SV=1 \n",
      "\n",
      "  Query_1    sp|Q8VCA0|S22AJ_MOUSE    4.76077e-42 \t 1    Solute carrier family 22 member 19 OS=Mus musculus GN=Slc22a19 PE=1 SV=1 \n",
      "\n",
      "  Query_1    sp|Q8VCA0|S22AJ_MOUSE    6.39257e-21 \t 2    Solute carrier family 22 member 19 OS=Mus musculus GN=Slc22a19 PE=1 SV=1 \n",
      "\n",
      "  Query_1    sp|Q8R0S9|S22AM_MOUSE    1.04333e-41 \t 1    Solute carrier family 22 member 22 OS=Mus musculus GN=Slc22a22 PE=1 SV=1 \n",
      "\n",
      "  Query_1    sp|Q8R0S9|S22AM_MOUSE    1.43928e-19 \t 2    Solute carrier family 22 member 22 OS=Mus musculus GN=Slc22a22 PE=1 SV=1 \n",
      "\n",
      "  Query_1    sp|O35956|S22A6_RAT    5.54738e-35 \t 1    Solute carrier family 22 member 6 OS=Rattus norvegicus GN=Slc22a6 PE=1 SV=1 \n",
      "\n",
      "  Query_1    sp|O35956|S22A6_RAT    1.9653e-24 \t 2    Solute carrier family 22 member 6 OS=Rattus norvegicus GN=Slc22a6 PE=1 SV=1 \n",
      "\n",
      "  Query_1    sp|Q66J52|S226B_XENLA    2.35073e-34 \t 1    Solute carrier family 22 member 6-B OS=Xenopus laevis GN=slc22a6-b PE=2 SV=1 \n",
      "\n",
      "  Query_1    sp|Q66J52|S226B_XENLA    7.47759e-22 \t 2    Solute carrier family 22 member 6-B OS=Xenopus laevis GN=slc22a6-b PE=2 SV=1 \n",
      "\n",
      "  Query_1    sp|Q4U2R8|S22A6_HUMAN    2.88427e-33 \t 1    Solute carrier family 22 member 6 OS=Homo sapiens GN=SLC22A6 PE=1 SV=1 \n",
      "\n",
      "  Query_1    sp|Q4U2R8|S22A6_HUMAN    3.99476e-25 \t 2    Solute carrier family 22 member 6 OS=Homo sapiens GN=SLC22A6 PE=1 SV=1 \n",
      "\n",
      "  Query_1    sp|Q66J54|S226A_XENLA    6.81013e-33 \t 1    Solute carrier family 22 member 6-A OS=Xenopus laevis GN=slc22a6-a PE=2 SV=1 \n",
      "\n",
      "  Query_1    sp|Q66J54|S226A_XENLA    8.12649e-24 \t 2    Solute carrier family 22 member 6-A OS=Xenopus laevis GN=slc22a6-a PE=2 SV=1 \n",
      "\n",
      "  Query_1    sp|Q5RCH6|S22A6_PONAB    1.66812e-32 \t 1    Solute carrier family 22 member 6 OS=Pongo abelii GN=SLC22A6 PE=2 SV=1 \n",
      "\n",
      "  Query_1    sp|Q5RCH6|S22A6_PONAB    3.33423e-25 \t 2    Solute carrier family 22 member 6 OS=Pongo abelii GN=SLC22A6 PE=2 SV=1 \n",
      "\n",
      "  Query_1    sp|Q4W8A3|S22A6_MACFA    5.67735e-32 \t 1    Solute carrier family 22 member 6 OS=Macaca fascicularis GN=SLC22A6 PE=1 SV=1 \n",
      "\n",
      "  Query_1    sp|Q4W8A3|S22A6_MACFA    3.30207e-25 \t 2    Solute carrier family 22 member 6 OS=Macaca fascicularis GN=SLC22A6 PE=1 SV=1 \n",
      "\n",
      "  Query_1    sp|Q8VC69|S22A6_MOUSE    7.98384e-32 \t 1    Solute carrier family 22 member 6 OS=Mus musculus GN=Slc22a6 PE=1 SV=1 \n",
      "\n",
      "  Query_1    sp|Q8VC69|S22A6_MOUSE    8.58509e-23 \t 2    Solute carrier family 22 member 6 OS=Mus musculus GN=Slc22a6 PE=1 SV=1 \n",
      "\n",
      "  Query_1    sp|Q6NYN7|S22A6_DANRE    1.28659e-30 \t 1    Solute carrier family 22 member 6 OS=Danio rerio GN=slc22a6 PE=2 SV=1 \n",
      "\n",
      "  Query_1    sp|Q6NYN7|S22A6_DANRE    7.18428e-22 \t 2    Solute carrier family 22 member 6 OS=Danio rerio GN=slc22a6 PE=2 SV=1 \n",
      "\n",
      "  Query_1    sp|Q5R9C4|S22A8_PONAB    1.87627e-30 \t 1    Solute carrier family 22 member 8 OS=Pongo abelii GN=SLC22A8 PE=2 SV=1 \n",
      "\n",
      "  Query_1    sp|Q5R9C4|S22A8_PONAB    2.20276e-22 \t 2    Solute carrier family 22 member 8 OS=Pongo abelii GN=SLC22A8 PE=2 SV=1 \n",
      "\n",
      "  Query_1    sp|Q9TSY7|S22A6_RABIT    3.07812e-30 \t 1    Solute carrier family 22 member 6 OS=Oryctolagus cuniculus GN=SLC22A6 PE=2 SV=1 \n",
      "\n",
      "  Query_1    sp|Q9TSY7|S22A6_RABIT    1.3715e-25 \t 2    Solute carrier family 22 member 6 OS=Oryctolagus cuniculus GN=SLC22A6 PE=2 SV=1 \n",
      "\n",
      "  Query_1    sp|Q4W8A2|S22A8_MACFA    8.12098e-30 \t 1    Solute carrier family 22 member 8 OS=Macaca fascicularis GN=SLC22A8 PE=2 SV=1 \n",
      "\n",
      "  Query_1    sp|Q4W8A2|S22A8_MACFA    3.38035e-23 \t 2    Solute carrier family 22 member 8 OS=Macaca fascicularis GN=SLC22A8 PE=2 SV=1 \n",
      "\n",
      "  Query_1    sp|Q864Z3|S22A6_BOVIN    9.46439e-30 \t 1    Solute carrier family 22 member 6 OS=Bos taurus GN=SLC22A6 PE=2 SV=2 \n",
      "\n",
      "  Query_1    sp|Q864Z3|S22A6_BOVIN    8.84813e-23 \t 2    Solute carrier family 22 member 6 OS=Bos taurus GN=SLC22A6 PE=2 SV=2 \n",
      "\n",
      "  Query_1    sp|A6NK97|S22AK_HUMAN    1.72054e-29 \t 1    Solute carrier family 22 member 20 OS=Homo sapiens GN=SLC22A20 PE=2 SV=1 \n",
      "\n",
      "  Query_1    sp|A6NK97|S22AK_HUMAN    4.39103e-24 \t 2    Solute carrier family 22 member 20 OS=Homo sapiens GN=SLC22A20 PE=2 SV=1 \n",
      "\n",
      "  Query_1    sp|Q8TCC7|S22A8_HUMAN    8.93333e-29 \t 1    Solute carrier family 22 member 8 OS=Homo sapiens GN=SLC22A8 PE=1 SV=1 \n",
      "\n",
      "  Query_1    sp|Q8TCC7|S22A8_HUMAN    3.02317e-22 \t 2    Solute carrier family 22 member 8 OS=Homo sapiens GN=SLC22A8 PE=1 SV=1 \n",
      "\n",
      "  Query_1    sp|Q80UJ1|S22AK_MOUSE    1.13083e-28 \t 1    Solute carrier family 22 member 20 OS=Mus musculus GN=Slc22a20 PE=1 SV=1 \n",
      "\n",
      "  Query_1    sp|Q80UJ1|S22AK_MOUSE    1.29389e-26 \t 2    Solute carrier family 22 member 20 OS=Mus musculus GN=Slc22a20 PE=1 SV=1 \n",
      "\n",
      "  Query_1    sp|Q70BM6|S22A8_PIG    5.43083e-28 \t 1    Solute carrier family 22 member 8 OS=Sus scrofa GN=SLC22A8 PE=2 SV=1 \n",
      "\n",
      "  Query_1    sp|Q70BM6|S22A8_PIG    1.32801e-22 \t 2    Solute carrier family 22 member 8 OS=Sus scrofa GN=SLC22A8 PE=2 SV=1 \n",
      "\n",
      "  Query_1    sp|O88909|S22A8_MOUSE    7.26336e-28 \t 1    Solute carrier family 22 member 8 OS=Mus musculus GN=Slc22a8 PE=1 SV=2 \n",
      "\n",
      "  Query_1    sp|O88909|S22A8_MOUSE    8.97063e-23 \t 2    Solute carrier family 22 member 8 OS=Mus musculus GN=Slc22a8 PE=1 SV=2 \n",
      "\n",
      "  Query_1    sp|O57379|S22A6_PSEAM    2.18954e-27 \t 1    Solute carrier family 22 member 6 OS=Pseudopleuronectes americanus GN=SLC22A6 PE=1 SV=1 \n",
      "\n",
      "  Query_1    sp|O57379|S22A6_PSEAM    2.06833e-19 \t 2    Solute carrier family 22 member 6 OS=Pseudopleuronectes americanus GN=SLC22A6 PE=1 SV=1 \n",
      "\n",
      "  Query_1    sp|Q9R1U7|S22A8_RAT    4.89152e-27 \t 1    Solute carrier family 22 member 8 OS=Rattus norvegicus GN=Slc22a8 PE=1 SV=1 \n",
      "\n",
      "  Query_1    sp|Q9R1U7|S22A8_RAT    3.92799e-22 \t 2    Solute carrier family 22 member 8 OS=Rattus norvegicus GN=Slc22a8 PE=1 SV=1 \n",
      "\n",
      "  Query_1    sp|Q8HY24|S22A8_RABIT    1.58173e-26 \t 1    Solute carrier family 22 member 8 OS=Oryctolagus cuniculus GN=SLC22A8 PE=2 SV=1 \n",
      "\n",
      "  Query_1    sp|Q8HY24|S22A8_RABIT    1.62805e-20 \t 2    Solute carrier family 22 member 8 OS=Oryctolagus cuniculus GN=SLC22A8 PE=2 SV=1 \n",
      "\n",
      "  Query_1    sp|Q8MK48|S22A6_PIG    3.11031e-25 \t 1    Solute carrier family 22 member 6 OS=Sus scrofa GN=SLC22A6 PE=1 SV=2 \n",
      "\n",
      "  Query_1    sp|Q8MK48|S22A6_PIG    8.65746e-24 \t 2    Solute carrier family 22 member 6 OS=Sus scrofa GN=SLC22A6 PE=1 SV=2 \n",
      "\n",
      "  Query_1    sp|A6QLW8|S22A7_BOVIN    4.80411e-23 \t 1    Solute carrier family 22 member 7 OS=Bos taurus GN=SLC22A7 PE=2 SV=1 \n",
      "\n",
      "  Query_1    sp|A6QLW8|S22A7_BOVIN    3.59154e-17 \t 2    Solute carrier family 22 member 7 OS=Bos taurus GN=SLC22A7 PE=2 SV=1 \n",
      "\n",
      "  Query_1    sp|Q9Y694|S22A7_HUMAN    5.10092e-22 \t 1    Solute carrier family 22 member 7 OS=Homo sapiens GN=SLC22A7 PE=1 SV=1 \n",
      "\n",
      "  Query_1    sp|Q9Y694|S22A7_HUMAN    2.89038e-15 \t 2    Solute carrier family 22 member 7 OS=Homo sapiens GN=SLC22A7 PE=1 SV=1 \n",
      "\n",
      "  Query_1    sp|Q1RPP5|S22A7_PIG    2.87209e-21 \t 1    Solute carrier family 22 member 7 OS=Sus scrofa GN=SLC22A7 PE=2 SV=1 \n",
      "\n",
      "  Query_1    sp|Q1RPP5|S22A7_PIG    1.02504e-17 \t 2    Solute carrier family 22 member 7 OS=Sus scrofa GN=SLC22A7 PE=2 SV=1 \n",
      "\n",
      "  Query_1    sp|Q5RLM2|S22A7_RAT    4.01662e-21 \t 1    Solute carrier family 22 member 7 OS=Rattus norvegicus GN=Slc22a7 PE=1 SV=1 \n",
      "\n",
      "  Query_1    sp|Q5RLM2|S22A7_RAT    1.46591e-16 \t 2    Solute carrier family 22 member 7 OS=Rattus norvegicus GN=Slc22a7 PE=1 SV=1 \n",
      "\n",
      "  Query_1    sp|Q5R540|S22A7_PONAB    5.45494e-21 \t 1    Solute carrier family 22 member 7 OS=Pongo abelii GN=SLC22A7 PE=2 SV=1 \n",
      "\n",
      "  Query_1    sp|Q5R540|S22A7_PONAB    1.13206e-15 \t 2    Solute carrier family 22 member 7 OS=Pongo abelii GN=SLC22A7 PE=2 SV=1 \n",
      "\n",
      "  Query_1    sp|Q91WU2|S22A7_MOUSE    3.67797e-20 \t 1    Solute carrier family 22 member 7 OS=Mus musculus GN=Slc22a7 PE=1 SV=1 \n",
      "\n",
      "  Query_1    sp|Q91WU2|S22A7_MOUSE    3.67157e-16 \t 2    Solute carrier family 22 member 7 OS=Mus musculus GN=Slc22a7 PE=1 SV=1 \n",
      "\n",
      "  Query_1    sp|Q3YAW7|S22A7_RABIT    1.70574e-19 \t 1    Solute carrier family 22 member 7 OS=Oryctolagus cuniculus GN=SLC22A7 PE=2 SV=1 \n",
      "\n",
      "  Query_1    sp|Q3YAW7|S22A7_RABIT    7.37303e-16 \t 2    Solute carrier family 22 member 7 OS=Oryctolagus cuniculus GN=SLC22A7 PE=2 SV=1 \n",
      "\n",
      "  Query_1    sp|Q9Z0E8|S22A5_MOUSE    5.39632e-17 \t 1    Solute carrier family 22 member 5 OS=Mus musculus GN=Slc22a5 PE=1 SV=1 \n",
      "\n",
      "  Query_1    sp|Q9WTN6|S22AL_MOUSE    2.46943e-16 \t 1    Solute carrier family 22 member 21 OS=Mus musculus GN=Slc22a21 PE=2 SV=1 \n",
      "\n",
      "  Query_1    sp|O70594|S22A5_RAT    4.44505e-15 \t 1    Solute carrier family 22 member 5 OS=Rattus norvegicus GN=Slc22a5 PE=1 SV=1 \n",
      "\n",
      "  Query_1    sp|Q9Z306|S22A4_MOUSE    1.7029e-14 \t 1    Solute carrier family 22 member 4 OS=Mus musculus GN=Slc22a4 PE=1 SV=1 \n",
      "\n",
      "  Query_1    sp|Q9Y226|S22AD_HUMAN    4.16716e-14 \t 1    Solute carrier family 22 member 13 OS=Homo sapiens GN=SLC22A13 PE=2 SV=2 \n",
      "\n",
      "  Query_1    sp|Q9R141|S22A4_RAT    8.25159e-14 \t 1    Solute carrier family 22 member 4 OS=Rattus norvegicus GN=Slc22a4 PE=1 SV=1 \n",
      "\n",
      "  Query_1    sp|Q9H015|S22A4_HUMAN    3.12319e-13 \t 1    Solute carrier family 22 member 4 OS=Homo sapiens GN=SLC22A4 PE=1 SV=3 \n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  Query_1    sp|A9CB25|S22A4_PAPAN    1.56162e-12 \t 1    Solute carrier family 22 member 4 OS=Papio anubis GN=SLC22A4 PE=3 SV=1 \n",
      "\n",
      "  Query_1    sp|O76082|S22A5_HUMAN    1.5994e-12 \t 1    Solute carrier family 22 member 5 OS=Homo sapiens GN=SLC22A5 PE=1 SV=1 \n",
      "\n",
      "  Query_1    sp|Q17QN9|S22AG_BOVIN    4.53963e-12 \t 1    Solute carrier family 22 member 16 OS=Bos taurus GN=SLC22A16 PE=2 SV=2 \n",
      "\n",
      "  Query_1    sp|Q6A4L0|S22AD_MOUSE    6.54368e-11 \t 1    Solute carrier family 22 member 13 OS=Mus musculus GN=Slc22a13 PE=1 SV=3 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "from Bio.Blast import NCBIXML\n",
    "\n",
    "filename = '../source/test.blastout.xml'\n",
    "e_threshold = 1e-10\n",
    "\n",
    "handle = open(filename)\n",
    "blast_records = NCBIXML.parse(handle)\n",
    "\n",
    "print('Query Name \\t Hit ID \\t\\t e-Value    Hsp Number \\t Hit Description \\n ----------------------------------------------------------------------------------------------------------------')\n",
    "for blast_record in blast_records:\n",
    "    for alignment in blast_record.alignments:\n",
    "        hsp_num = 0\n",
    "        for hsp in alignment.hsps:\n",
    "            hsp_num += 1\n",
    "            if hsp.expect < e_threshold:\n",
    "                print(' ', blast_record.query_id, '  ', alignment.hit_id, '  ', hsp.expect, '\\t', hsp_num, '  ', alignment.hit_def, '\\n')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
