{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Tarea 4: *ExPASy*\n",
    "\n",
    "***Paulina Rosales Becerra***"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='indice'></a>\n",
    "## Índice\n",
    "1. [Funciones](#0)\n",
    "1. [Ejercicio 1](#1)\n",
    "1. [Ejercicio 2](#2)\n",
    "1. [Ejercicio 3](#3)\n",
    "\n",
    "**Nota:** La mayoría de los códigos cuentan con variables *filename* y *structure_id* para poder editar la ruta del archivo *input* o *output* en caso de ser requerido."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='0'></a>\n",
    "## Funciones\n",
    "[Regresar al índice](#indice)\n",
    "\n",
    "Utilizadas para evitar errores producidos por la biblioteca SwissProt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Copyright 2007 by Michiel de Hoon.  All rights reserved.\n",
    "# This code is part of the Biopython distribution and governed by its\n",
    "# license.  Please see the LICENSE file that should have been included\n",
    "# as part of this package.\n",
    "\"\"\"Code to work with the sprotXX.dat file from SwissProt.\n",
    "\n",
    "https://web.expasy.org/docs/userman.html\n",
    "\n",
    "Classes:\n",
    " - Record             Holds SwissProt data.\n",
    " - Reference          Holds reference data from a SwissProt record.\n",
    "\n",
    "Functions:\n",
    " - read               Read one SwissProt record\n",
    " - parse              Read multiple SwissProt records\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "\n",
    "import io\n",
    "\n",
    "from Bio.SeqFeature import (\n",
    "    SeqFeature,\n",
    "    FeatureLocation,\n",
    "    ExactPosition,\n",
    "    BeforePosition,\n",
    "    AfterPosition,\n",
    "    UncertainPosition,\n",
    "    UnknownPosition,\n",
    ")\n",
    "\n",
    "\n",
    "class SwissProtParserError(ValueError):\n",
    "    \"\"\"An error occurred while parsing a SwissProt file.\"\"\"\n",
    "\n",
    "    def __init__(self, *args, line=None):\n",
    "        \"\"\"Create a SwissProtParserError object with the offending line.\"\"\"\n",
    "        super().__init__(*args)\n",
    "        self.line = line\n",
    "\n",
    "\n",
    "class Record:\n",
    "    \"\"\"Holds information from a SwissProt record.\n",
    "\n",
    "    Attributes:\n",
    "     - entry_name        Name of this entry, e.g. RL1_ECOLI.\n",
    "     - data_class        Either 'STANDARD' or 'PRELIMINARY'.\n",
    "     - molecule_type     Type of molecule, 'PRT',\n",
    "     - sequence_length   Number of residues.\n",
    "     - accessions        List of the accession numbers, e.g. ['P00321']\n",
    "     - created           A tuple of (date, release).\n",
    "     - sequence_update   A tuple of (date, release).\n",
    "     - annotation_update A tuple of (date, release).\n",
    "     - description       Free-format description.\n",
    "     - gene_name         Gene name.  See userman.txt for description.\n",
    "     - organism          The source of the sequence.\n",
    "     - organelle         The origin of the sequence.\n",
    "     - organism_classification  The taxonomy classification.  List of strings.\n",
    "       (http://www.ncbi.nlm.nih.gov/Taxonomy/)\n",
    "     - taxonomy_id       A list of NCBI taxonomy id's.\n",
    "     - host_organism     A list of names of the hosts of a virus, if any.\n",
    "     - host_taxonomy_id  A list of NCBI taxonomy id's of the hosts, if any.\n",
    "     - references        List of Reference objects.\n",
    "     - comments          List of strings.\n",
    "     - cross_references  List of tuples (db, id1[, id2][, id3]).  See the docs.\n",
    "     - keywords          List of the keywords.\n",
    "     - features          List of tuples (key name, from, to, description).\n",
    "       from and to can be either integers for the residue\n",
    "       numbers, '<', '>', or '?'\n",
    "     - protein_existence Numerical value describing the evidence for the existence of the protein.\n",
    "     - seqinfo           tuple of (length, molecular weight, CRC32 value)\n",
    "     - sequence          The sequence.\n",
    "\n",
    "    Examples\n",
    "    --------\n",
    "    >>> from Bio import SwissProt\n",
    "    >>> example_filename = \"SwissProt/sp008\"\n",
    "    >>> with open(example_filename) as handle:\n",
    "    ...     records = SwissProt.parse(handle)\n",
    "    ...     for record in records:\n",
    "    ...         print(record.entry_name)\n",
    "    ...         print(\",\".join(record.accessions))\n",
    "    ...         print(record.keywords)\n",
    "    ...         print(repr(record.organism))\n",
    "    ...         print(record.sequence[:20] + \"...\")\n",
    "    ...\n",
    "    1A02_HUMAN\n",
    "    P01892,P06338,P30514,P30444,P30445,P30446,Q29680,Q29899,Q95352,Q29837,Q95380\n",
    "    ['MHC I', 'Transmembrane', 'Glycoprotein', 'Signal', 'Polymorphism', '3D-structure']\n",
    "    'Homo sapiens (Human).'\n",
    "    MAVMAPRTLVLLLSGALALT...\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "    def __init__(self):\n",
    "        \"\"\"Initialize the class.\"\"\"\n",
    "        self.entry_name = None\n",
    "        self.data_class = None\n",
    "        self.molecule_type = None\n",
    "        self.sequence_length = None\n",
    "\n",
    "        self.accessions = []\n",
    "        self.created = None\n",
    "        self.sequence_update = None\n",
    "        self.annotation_update = None\n",
    "\n",
    "        self.description = []\n",
    "        self.gene_name = \"\"\n",
    "        self.organism = []\n",
    "        self.organelle = \"\"\n",
    "        self.organism_classification = []\n",
    "        self.taxonomy_id = []\n",
    "        self.host_organism = []\n",
    "        self.host_taxonomy_id = []\n",
    "        self.references = []\n",
    "        self.comments = []\n",
    "        self.cross_references = []\n",
    "        self.keywords = []\n",
    "        self.features = []\n",
    "        self.protein_existence = \"\"\n",
    "\n",
    "        self.seqinfo = None\n",
    "        self.sequence = \"\"\n",
    "\n",
    "\n",
    "class Reference:\n",
    "    \"\"\"Holds information from one reference in a SwissProt entry.\n",
    "\n",
    "    Attributes:\n",
    "     - number      Number of reference in an entry.\n",
    "     - evidence    Evidence code.  List of strings.\n",
    "     - positions   Describes extent of work.  List of strings.\n",
    "     - comments    Comments.  List of (token, text).\n",
    "     - references  References.  List of (dbname, identifier).\n",
    "     - authors     The authors of the work.\n",
    "     - title       Title of the work.\n",
    "     - location    A citation for the work.\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "    def __init__(self):\n",
    "        \"\"\"Initialize the class.\"\"\"\n",
    "        self.number = None\n",
    "        self.positions = []\n",
    "        self.comments = []\n",
    "        self.references = []\n",
    "        self.authors = []\n",
    "        self.title = []\n",
    "        self.location = []\n",
    "\n",
    "\n",
    "class FeatureTable(SeqFeature):\n",
    "    \"\"\"Stores feature annotations for specific regions of the sequence.\n",
    "\n",
    "    This is a subclass of SeqFeature, defined in Bio.SeqFeature, where the\n",
    "    attributes are used as follows:\n",
    "\n",
    "     - ``location``: location of the feature on the canonical or isoform\n",
    "       sequence; the location is stored as an instance of FeatureLocation,\n",
    "       defined in Bio.SeqFeature, with the ref attribute set to the isoform\n",
    "       ID referring to the canonical or isoform sequence on which the feature\n",
    "       is defined\n",
    "     - ``id``: unique and stable identifier (FTId), only provided for features\n",
    "       belonging to the types CARBOHYD, CHAIN, PEPTIDE, PROPEP, VARIANT, or\n",
    "       VAR_SEQ\n",
    "     - ``type``: indicates the type of feature, as defined by the UniProt\n",
    "       Knowledgebase documentation:\n",
    "\n",
    "        - ACT_SITE: amino acid(s) involved in the activity of an enzyme\n",
    "        - BINDING:  binding site for any chemical group\n",
    "        - CARBOHYD: glycosylation site; an FTId identifier to the GlyConnect\n",
    "          database is provided if annotated there\n",
    "        - CA_BIND:  calcium-binding region\n",
    "        - CHAIN:    polypeptide chain in the mature protein\n",
    "        - COILED:   coiled-coil region\n",
    "        - COMPBIAS: compositionally biased region\n",
    "        - CONFLICT: different sources report differing sequences\n",
    "        - CROSSLNK: posttransationally formed amino acid bond\n",
    "        - DISULFID: disulfide bond\n",
    "        - DNA_BIND: DNA-binding region\n",
    "        - DOMAIN:   domain, defined as a specific combination of secondary\n",
    "          structures organized into a characteristic three-dimensional\n",
    "          structure or fold\n",
    "        - INIT_MET: initiator methionine\n",
    "        - INTRAMEM: region located in a membrane without crossing it\n",
    "        - HELIX:    alpha-, 3(10)-, or pi-helix secondary structure\n",
    "        - LIPID:    covalent binding of a lipid moiety\n",
    "        - METAL:    binding site for a metal ion\n",
    "        - MOD_RES:  posttranslational modification (PTM) of a residue,\n",
    "          annotated by the controlled vocabulary defined by the ptmlist.txt\n",
    "          document on the UniProt website\n",
    "        - MOTIF:    short sequence motif of biological interest\n",
    "        - MUTAGEN:  site experimentally altered by mutagenesis\n",
    "        - NON_CONS: non-consecutive residues\n",
    "        - NON_STD:  non-standard amino acid\n",
    "        - NON_TER:  the residue at an extremity of the sequence is not the\n",
    "          terminal residue\n",
    "        - NP_BIND:  nucleotide phosphate-binding region\n",
    "        - PEPTIDE:  released active mature polypeptide\n",
    "        - PROPEP:   any processed propeptide\n",
    "        - REGION:   region of interest in the sequence\n",
    "        - REPEAT:   internal sequence repetition\n",
    "        - SIGNAL:   signal sequence (prepeptide)\n",
    "        - SITE:     amino-acid site of interest not represented by another\n",
    "          feature key\n",
    "        - STRAND:   beta-strand secondary structure; either a hydrogen-bonded\n",
    "          extended beta strand or a residue in an isolated beta-bridge\n",
    "        - TOPO_DOM: topological domain\n",
    "        - TRANSIT:  transit peptide (mitochondrion, chloroplast, thylakoid,\n",
    "          cyanelle, peroxisome, etc.)\n",
    "        - TRANSMEM: transmembrane region\n",
    "        - TURN:     H-bonded turn (3-, 4-, or 5-turn)\n",
    "        - UNSURE:   uncertainties in the sequence\n",
    "        - VARIANT:  sequence variant; an FTId is provided for protein sequence\n",
    "          variants of Hominidae (great apes and humans)\n",
    "        - VAR_SEQ:  sequence variant produced by alternative splicing,\n",
    "          alternative promoter usage, alternative initiation, or ribosomal\n",
    "          frameshifting\n",
    "        - ZN_FING:  zinc finger region\n",
    "\n",
    "     - qualifiers   A dictionary of additional information, which may include\n",
    "       the feature evidence and free-text notes. While SwissProt includes the\n",
    "       feature identifier code (FTId) as a qualifier, it is stored as the\n",
    "       attribute ID of the FeatureTable object.\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "\n",
    "def parse(source):\n",
    "    \"\"\"Read multiple SwissProt records from file.\n",
    "\n",
    "    Argument source is a file-like object or a path to a file.\n",
    "\n",
    "    Returns a generator object which yields Bio.SwissProt.Record() objects.\n",
    "    \"\"\"\n",
    "    handle = _open(source)\n",
    "    try:\n",
    "        while True:\n",
    "            record = _read(handle)\n",
    "            if not record:\n",
    "                return\n",
    "            yield record\n",
    "    finally:\n",
    "        if handle is not source:\n",
    "            handle.close()\n",
    "\n",
    "\n",
    "def read(source):\n",
    "    \"\"\"Read one SwissProt record from file.\n",
    "\n",
    "    Argument source is a file-like object or a path to a file.\n",
    "\n",
    "    Returns a Record() object.\n",
    "    \"\"\"\n",
    "    handle = _open(source)\n",
    "    try:\n",
    "        record = _read(handle)\n",
    "        if not record:\n",
    "            raise ValueError(\"No SwissProt record found\")\n",
    "        # We should have reached the end of the record by now.\n",
    "        # Try to read one more line to be sure:\n",
    "        try:\n",
    "            next(handle)\n",
    "        except StopIteration:\n",
    "            return record\n",
    "        raise ValueError(\"More than one SwissProt record found\")\n",
    "    finally:\n",
    "        if handle is not source:\n",
    "            handle.close()\n",
    "\n",
    "\n",
    "# Everything below is considered private\n",
    "\n",
    "\n",
    "def _open(source):\n",
    "    try:\n",
    "        handle = open(source)\n",
    "        return handle\n",
    "    except TypeError:\n",
    "        handle = source\n",
    "        if handle.read(0) == \"\":\n",
    "            # handle is text; assume the encoding is compatible with ASCII\n",
    "            return handle\n",
    "        # handle is binary; SwissProt encoding is always ASCII\n",
    "        return io.TextIOWrapper(handle, encoding=\"ASCII\")\n",
    "\n",
    "\n",
    "def _read(handle):\n",
    "    record = None\n",
    "    unread = \"\"\n",
    "    try:\n",
    "        line = next(handle)\n",
    "    except StopIteration:\n",
    "        return record\n",
    "    key, value = line[:2], line[5:].rstrip()\n",
    "    if key != \"ID\":\n",
    "        raise SwissProtParserError(\"Failed to find ID in first line\", line=line)\n",
    "    record = Record()\n",
    "    _read_id(record, line)\n",
    "    _sequence_lines = []\n",
    "    for line in handle:\n",
    "        key, value = line[:2], line[5:].rstrip()\n",
    "        if unread:\n",
    "            value = unread + \" \" + value\n",
    "            unread = \"\"\n",
    "        if key == \"AC\":\n",
    "            accessions = value.rstrip(\";\").split(\"; \")\n",
    "            record.accessions.extend(accessions)\n",
    "        elif key == \"DT\":\n",
    "            _read_dt(record, line)\n",
    "        elif key == \"DE\":\n",
    "            record.description.append(value.strip())\n",
    "        elif key == \"GN\":\n",
    "            if record.gene_name:\n",
    "                record.gene_name += \" \"\n",
    "            record.gene_name += value\n",
    "        elif key == \"OS\":\n",
    "            record.organism.append(value)\n",
    "        elif key == \"OG\":\n",
    "            record.organelle += line[5:]\n",
    "        elif key == \"OC\":\n",
    "            cols = value.rstrip(\";.\").split(\"; \")\n",
    "            record.organism_classification.extend(cols)\n",
    "        elif key == \"OX\":\n",
    "            _read_ox(record, line)\n",
    "        elif key == \"OH\":\n",
    "            _read_oh(record, line)\n",
    "        elif key == \"RN\":\n",
    "            reference = Reference()\n",
    "            _read_rn(reference, value)\n",
    "            record.references.append(reference)\n",
    "        elif key == \"RP\":\n",
    "            assert record.references, \"RP: missing RN\"\n",
    "            record.references[-1].positions.append(value)\n",
    "        elif key == \"RC\":\n",
    "            assert record.references, \"RC: missing RN\"\n",
    "            reference = record.references[-1]\n",
    "            unread = _read_rc(reference, value)\n",
    "        elif key == \"RX\":\n",
    "            assert record.references, \"RX: missing RN\"\n",
    "            reference = record.references[-1]\n",
    "            _read_rx(reference, value)\n",
    "        elif key == \"RL\":\n",
    "            assert record.references, \"RL: missing RN\"\n",
    "            reference = record.references[-1]\n",
    "            reference.location.append(value)\n",
    "        # In UniProt release 1.12 of 6/21/04, there is a new RG\n",
    "        # (Reference Group) line, which references a group instead of\n",
    "        # an author.  Each block must have at least 1 RA or RG line.\n",
    "        elif key == \"RA\":\n",
    "            assert record.references, \"RA: missing RN\"\n",
    "            reference = record.references[-1]\n",
    "            reference.authors.append(value)\n",
    "        elif key == \"RG\":\n",
    "            assert record.references, \"RG: missing RN\"\n",
    "            reference = record.references[-1]\n",
    "            reference.authors.append(value)\n",
    "        elif key == \"RT\":\n",
    "            assert record.references, \"RT: missing RN\"\n",
    "            reference = record.references[-1]\n",
    "            reference.title.append(value)\n",
    "        elif key == \"CC\":\n",
    "            _read_cc(record, line)\n",
    "        elif key == \"DR\":\n",
    "            _read_dr(record, value)\n",
    "        elif key == \"PE\":\n",
    "            _read_pe(record, value)\n",
    "        elif key == \"KW\":\n",
    "            _read_kw(record, value)\n",
    "        elif key == \"FT\":\n",
    "            _read_ft(record, line)\n",
    "        elif key == \"SQ\":\n",
    "            cols = value.split()\n",
    "            assert len(cols) == 7, \"I don't understand SQ line %s\" % line\n",
    "            # Do more checking here?\n",
    "            record.seqinfo = int(cols[1]), int(cols[3]), cols[5]\n",
    "        elif key == \"  \":\n",
    "            _sequence_lines.append(value.replace(\" \", \"\").rstrip())\n",
    "        elif key == \"//\":\n",
    "            # Join multiline data into one string\n",
    "            record.description = \" \".join(record.description)\n",
    "            record.organism = \" \".join(record.organism)\n",
    "            record.organelle = record.organelle.rstrip()\n",
    "            for reference in record.references:\n",
    "                reference.authors = \" \".join(reference.authors).rstrip(\";\")\n",
    "                if reference.title:\n",
    "                    title = reference.title[0]\n",
    "                    for fragment in reference.title[1:]:\n",
    "                        if not title.endswith(\"-\"):\n",
    "                            title += \" \"\n",
    "                        title += fragment\n",
    "                    title = title.rstrip(\";\")\n",
    "                    if title.startswith('\"') and title.endswith('\"'):\n",
    "                        title = title[1:-1]  # remove quotes\n",
    "                else:\n",
    "                    title = \"\"\n",
    "                reference.title = title\n",
    "                reference.location = \" \".join(reference.location)\n",
    "            record.sequence = \"\".join(_sequence_lines)\n",
    "            return record\n",
    "        elif key == \"**\":\n",
    "            # Do this one last, as it will almost never occur.\n",
    "            # See Bug 2353, some files from the EBI have extra lines\n",
    "            # starting \"**\" (two asterisks/stars).  They appear\n",
    "            # to be unofficial automated annotations. e.g.\n",
    "            # **\n",
    "            # **   #################    INTERNAL SECTION    ##################\n",
    "            # **HA SAM; Annotated by PicoHamap 1.88; MF_01138.1; 09-NOV-2003.\n",
    "            pass\n",
    "        else:\n",
    "            raise SwissProtParserError(\"Unknown keyword '%s' found\" % key, line=line)\n",
    "    if record:\n",
    "        raise ValueError(\"Unexpected end of stream.\")\n",
    "\n",
    "\n",
    "def _read_id(record, line):\n",
    "    cols = line[5:].split()\n",
    "    # Prior to release 51, included with MoleculeType:\n",
    "    # ID   EntryName DataClass; MoleculeType; SequenceLength AA.\n",
    "    #\n",
    "    # Newer files lack the MoleculeType:\n",
    "    # ID   EntryName DataClass; SequenceLength AA.\n",
    "    if len(cols) == 5:\n",
    "        record.entry_name = cols[0]\n",
    "        record.data_class = cols[1].rstrip(\";\")\n",
    "        record.molecule_type = cols[2].rstrip(\";\")\n",
    "        record.sequence_length = int(cols[3])\n",
    "    elif len(cols) == 4:\n",
    "        record.entry_name = cols[0]\n",
    "        record.data_class = cols[1].rstrip(\";\")\n",
    "        record.molecule_type = None\n",
    "        record.sequence_length = int(cols[2])\n",
    "    else:\n",
    "        raise SwissProtParserError(\"ID line has unrecognised format\", line=line)\n",
    "    # check if the data class is one of the allowed values\n",
    "    allowed = (\"STANDARD\", \"PRELIMINARY\", \"IPI\", \"Reviewed\", \"Unreviewed\")\n",
    "    if record.data_class not in allowed:\n",
    "        message = \"Unrecognized data class '%s'\" % record.data_class\n",
    "        raise SwissProtParserError(message, line=line)\n",
    "\n",
    "    # molecule_type should be 'PRT' for PRoTein\n",
    "    # Note that has been removed in recent releases (set to None)\n",
    "    if record.molecule_type not in (None, \"PRT\"):\n",
    "        message = \"Unrecognized molecule type '%s'\" % record.molecule_type\n",
    "        raise SwissProtParserError(message, line=line)\n",
    "\n",
    "\n",
    "def _read_dt(record, line):\n",
    "    value = line[5:]\n",
    "    uprline = value.upper()\n",
    "    cols = value.rstrip().split()\n",
    "    if (\n",
    "        \"CREATED\" in uprline\n",
    "        or \"LAST SEQUENCE UPDATE\" in uprline\n",
    "        or \"LAST ANNOTATION UPDATE\" in uprline\n",
    "    ):\n",
    "        # Old style DT line\n",
    "        # =================\n",
    "        # e.g.\n",
    "        # DT   01-FEB-1995 (Rel. 31, Created)\n",
    "        # DT   01-FEB-1995 (Rel. 31, Last sequence update)\n",
    "        # DT   01-OCT-2000 (Rel. 40, Last annotation update)\n",
    "        #\n",
    "        # or:\n",
    "        # DT   08-JAN-2002 (IPI Human rel. 2.3, Created)\n",
    "        # ...\n",
    "\n",
    "        # find where the version information will be located\n",
    "        # This is needed for when you have cases like IPI where\n",
    "        # the release version is in a different spot:\n",
    "        # DT   08-JAN-2002 (IPI Human rel. 2.3, Created)\n",
    "        uprcols = uprline.split()\n",
    "        rel_index = -1\n",
    "        for index in range(len(uprcols)):\n",
    "            if \"REL.\" in uprcols[index]:\n",
    "                rel_index = index\n",
    "        assert rel_index >= 0, \"Could not find Rel. in DT line: %s\" % line\n",
    "        version_index = rel_index + 1\n",
    "        # get the version information\n",
    "        str_version = cols[version_index].rstrip(\",\")\n",
    "        # no version number\n",
    "        if str_version == \"\":\n",
    "            version = 0\n",
    "        # dot versioned\n",
    "        elif \".\" in str_version:\n",
    "            version = str_version\n",
    "        # integer versioned\n",
    "        else:\n",
    "            version = int(str_version)\n",
    "        date = cols[0]\n",
    "\n",
    "        if \"CREATED\" in uprline:\n",
    "            record.created = date, version\n",
    "        elif \"LAST SEQUENCE UPDATE\" in uprline:\n",
    "            record.sequence_update = date, version\n",
    "        elif \"LAST ANNOTATION UPDATE\" in uprline:\n",
    "            record.annotation_update = date, version\n",
    "        else:\n",
    "            raise SwissProtParserError(\"Unrecognised DT (DaTe) line\", line=line)\n",
    "    elif (\n",
    "        \"INTEGRATED INTO\" in uprline\n",
    "        or \"SEQUENCE VERSION\" in uprline\n",
    "        or \"ENTRY VERSION\" in uprline\n",
    "    ):\n",
    "        # New style DT line\n",
    "        # =================\n",
    "        # As of UniProt Knowledgebase release 7.0 (including\n",
    "        # Swiss-Prot release 49.0 and TrEMBL release 32.0) the\n",
    "        # format of the DT lines and the version information\n",
    "        # in them was changed - the release number was dropped.\n",
    "        #\n",
    "        # For more information see bug 1948 and\n",
    "        # http://ca.expasy.org/sprot/relnotes/sp_news.html#rel7.0\n",
    "        #\n",
    "        # e.g.\n",
    "        # DT   01-JAN-1998, integrated into UniProtKB/Swiss-Prot.\n",
    "        # DT   15-OCT-2001, sequence version 3.\n",
    "        # DT   01-APR-2004, entry version 14.\n",
    "        #\n",
    "        # This is a new style DT line...\n",
    "\n",
    "        # The date should be in string cols[1]\n",
    "        # Get the version number if there is one.\n",
    "        # For the three DT lines above: 0, 3, 14\n",
    "        try:\n",
    "            version = 0\n",
    "            for s in cols[-1].split(\".\"):\n",
    "                if s.isdigit():\n",
    "                    version = int(s)\n",
    "        except ValueError:\n",
    "            version = 0\n",
    "        date = cols[0].rstrip(\",\")\n",
    "\n",
    "        # Re-use the historical property names, even though\n",
    "        # the meaning has changed slighty:\n",
    "        if \"INTEGRATED\" in uprline:\n",
    "            record.created = date, version\n",
    "        elif \"SEQUENCE VERSION\" in uprline:\n",
    "            record.sequence_update = date, version\n",
    "        elif \"ENTRY VERSION\" in uprline:\n",
    "            record.annotation_update = date, version\n",
    "        else:\n",
    "            raise SwissProtParserError(\"Unrecognised DT (DaTe) line\", line=line)\n",
    "    else:\n",
    "        raise SwissProtParserError(\"Failed to parse DT (DaTe) line\", line=line)\n",
    "\n",
    "\n",
    "def _read_ox(record, line):\n",
    "    # The OX line used to be in the simple format:\n",
    "    # OX   DESCRIPTION=ID[, ID]...;\n",
    "    # If there are too many id's to fit onto a line, then the ID's\n",
    "    # continue directly onto the next line, e.g.\n",
    "    # OX   DESCRIPTION=ID[, ID]...\n",
    "    # OX   ID[, ID]...;\n",
    "    # Currently, the description is always \"NCBI_TaxID\".\n",
    "    # To parse this, I need to check to see whether I'm at the\n",
    "    # first line.  If I am, grab the description and make sure\n",
    "    # it's an NCBI ID.  Then, grab all the id's.\n",
    "    #\n",
    "    # As of the 2014-10-01 release, there may be an evidence code, e.g.\n",
    "    # OX   NCBI_TaxID=418404 {ECO:0000313|EMBL:AEX14553.1};\n",
    "    # In the short term, we will ignore any evidence codes:\n",
    "    line = line.split(\"{\")[0]\n",
    "    if record.taxonomy_id:\n",
    "        ids = line[5:].rstrip().rstrip(\";\")\n",
    "    else:\n",
    "        descr, ids = line[5:].rstrip().rstrip(\";\").split(\"=\")\n",
    "        assert descr == \"NCBI_TaxID\", \"Unexpected taxonomy type %s\" % descr\n",
    "    record.taxonomy_id.extend(ids.split(\", \"))\n",
    "\n",
    "\n",
    "def _read_oh(record, line):\n",
    "    # Line type OH (Organism Host) for viral hosts\n",
    "    assert line[5:].startswith(\"NCBI_TaxID=\"), \"Unexpected %s\" % line\n",
    "    line = line[16:].rstrip()\n",
    "    assert line[-1] == \".\" and line.count(\";\") == 1, line\n",
    "    taxid, name = line[:-1].split(\";\")\n",
    "    record.host_taxonomy_id.append(taxid.strip())\n",
    "    record.host_organism.append(name.strip())\n",
    "\n",
    "\n",
    "def _read_rn(reference, rn):\n",
    "    # This used to be a very simple line with a reference number, e.g.\n",
    "    # RN   [1]\n",
    "    # As of the 2014-10-01 release, there may be an evidence code, e.g.\n",
    "    # RN   [1] {ECO:0000313|EMBL:AEX14553.1}\n",
    "    words = rn.split(None, 1)\n",
    "    number = words[0]\n",
    "    assert number.startswith(\"[\") and number.endswith(\"]\"), (\n",
    "        \"Missing brackets %s\" % number\n",
    "    )\n",
    "    reference.number = int(number[1:-1])\n",
    "    if len(words) > 1:\n",
    "        evidence = words[1]\n",
    "        assert evidence.startswith(\"{\") and evidence.endswith(\"}\"), (\n",
    "            \"Missing braces %s\" % evidence\n",
    "        )\n",
    "        reference.evidence = evidence[1:-1].split(\"|\")\n",
    "\n",
    "\n",
    "def _read_rc(reference, value):\n",
    "    cols = value.split(\";\")\n",
    "    if value[-1] == \";\":\n",
    "        unread = \"\"\n",
    "    else:\n",
    "        cols, unread = cols[:-1], cols[-1]\n",
    "    for col in cols:\n",
    "        if not col:  # last column will be the empty string\n",
    "            return\n",
    "        # The token is everything before the first '=' character.\n",
    "        i = col.find(\"=\")\n",
    "        if i >= 0:\n",
    "            token, text = col[:i], col[i + 1 :]\n",
    "            comment = token.lstrip(), text\n",
    "            reference.comments.append(comment)\n",
    "        else:\n",
    "            comment = reference.comments[-1]\n",
    "            comment = \"%s %s\" % (comment, col)\n",
    "            reference.comments[-1] = comment\n",
    "    return unread\n",
    "\n",
    "\n",
    "def _read_rx(reference, value):\n",
    "    # The basic (older?) RX line is of the form:\n",
    "    # RX   MEDLINE; 85132727.\n",
    "    # but there are variants of this that need to be dealt with (see below)\n",
    "\n",
    "    # CLD1_HUMAN in Release 39 and DADR_DIDMA in Release 33\n",
    "    # have extraneous information in the RX line.  Check for\n",
    "    # this and chop it out of the line.\n",
    "    # (noticed by katel@worldpath.net)\n",
    "    value = value.replace(\" [NCBI, ExPASy, Israel, Japan]\", \"\")\n",
    "\n",
    "    # RX lines can also be used of the form\n",
    "    # RX   PubMed=9603189;\n",
    "    # reported by edvard@farmasi.uit.no\n",
    "    # and these can be more complicated like:\n",
    "    # RX   MEDLINE=95385798; PubMed=7656980;\n",
    "    # RX   PubMed=15060122; DOI=10.1136/jmg 2003.012781;\n",
    "    # We look for these cases first and deal with them\n",
    "    warn = False\n",
    "    if \"=\" in value:\n",
    "        cols = value.split(\"; \")\n",
    "        cols = [x.strip() for x in cols]\n",
    "        cols = [x for x in cols if x]\n",
    "        for col in cols:\n",
    "            x = col.split(\"=\")\n",
    "            if len(x) != 2 or x == (\"DOI\", \"DOI\"):\n",
    "                warn = True\n",
    "                break\n",
    "            assert len(x) == 2, \"I don't understand RX line %s\" % value\n",
    "            reference.references.append((x[0], x[1].rstrip(\";\")))\n",
    "    # otherwise we assume we have the type 'RX   MEDLINE; 85132727.'\n",
    "    else:\n",
    "        cols = value.split(\"; \")\n",
    "        # normally we split into the three parts\n",
    "        if len(cols) != 2:\n",
    "            warn = True\n",
    "        else:\n",
    "            reference.references.append((cols[0].rstrip(\";\"), cols[1].rstrip(\".\")))\n",
    "    if warn:\n",
    "        import warnings\n",
    "        from Bio import BiopythonParserWarning\n",
    "\n",
    "        warnings.warn(\"Possibly corrupt RX line %r\" % value, BiopythonParserWarning)\n",
    "\n",
    "\n",
    "def _read_cc(record, line):\n",
    "    key, value = line[5:8], line[9:].rstrip()\n",
    "    if key == \"-!-\":  # Make a new comment\n",
    "        record.comments.append(value)\n",
    "    elif key == \"   \":  # add to the previous comment\n",
    "        if not record.comments:\n",
    "            # TCMO_STRGA in Release 37 has comment with no topic\n",
    "            record.comments.append(value)\n",
    "        else:\n",
    "            record.comments[-1] += \" \" + value\n",
    "\n",
    "\n",
    "def _read_dr(record, value):\n",
    "    cols = value.rstrip(\".\").split(\"; \")\n",
    "    record.cross_references.append(tuple(cols))\n",
    "\n",
    "\n",
    "def _read_pe(record, value):\n",
    "    pe = value.split(\":\")\n",
    "    record.protein_existence = int(pe[0])\n",
    "\n",
    "\n",
    "def _read_kw(record, value):\n",
    "    # Old style - semi-colon separated, multi-line. e.g. Q13639.txt\n",
    "    # KW   Alternative splicing; Cell membrane; Complete proteome;\n",
    "    # KW   Disulfide bond; Endosome; G-protein coupled receptor; Glycoprotein;\n",
    "    # KW   Lipoprotein; Membrane; Palmitate; Polymorphism; Receptor; Transducer;\n",
    "    # KW   Transmembrane.\n",
    "    #\n",
    "    # New style as of 2014-10-01 release with evidence codes, e.g. H2CNN8.txt\n",
    "    # KW   Monooxygenase {ECO:0000313|EMBL:AEX14553.1};\n",
    "    # KW   Oxidoreductase {ECO:0000313|EMBL:AEX14553.1}.\n",
    "    # For now to match the XML parser, drop the evidence codes.\n",
    "    for value in value.rstrip(\";.\").split(\"; \"):\n",
    "        if value.endswith(\"}\"):\n",
    "            # Discard the evidence code\n",
    "            value = value.rsplit(\"{\", 1)[0]\n",
    "        record.keywords.append(value.strip())\n",
    "\n",
    "\n",
    "def _read_ft(record, line):\n",
    "    name = line[5:13].rstrip()\n",
    "    if name:\n",
    "        if line[13:21] == \"        \":  # new-style FT line\n",
    "            location = line[21:80].rstrip()\n",
    "            try:\n",
    "                isoform_id, location = location.split(\":\")\n",
    "            except ValueError:\n",
    "                isoform_id = None\n",
    "            try:\n",
    "                from_res, to_res = location.split(\"..\")\n",
    "            except ValueError:\n",
    "                from_res = location\n",
    "                to_res = \"\"\n",
    "            qualifiers = {}\n",
    "        else:  # old-style FT line\n",
    "            from_res = line[14:20].lstrip()\n",
    "            to_res = line[21:27].lstrip()\n",
    "            isoform_id = None\n",
    "            description = line[34:75].rstrip()\n",
    "            qualifiers = {\"description\": description}\n",
    "        if from_res == \"?\":\n",
    "            from_res = UnknownPosition()\n",
    "        elif from_res.startswith(\"?\"):\n",
    "            position = int(from_res[1:]) - 1  # Python zero-based counting\n",
    "            from_res = UncertainPosition(position)\n",
    "        elif from_res.startswith(\"<\"):\n",
    "            position = int(from_res[1:]) - 1  # Python zero-based counting\n",
    "            from_res = BeforePosition(position)\n",
    "        else:\n",
    "            position = int(from_res) - 1  # Python zero-based counting\n",
    "            from_res = ExactPosition(position)\n",
    "        if to_res == \"\":\n",
    "            position = from_res + 1\n",
    "            to_res = ExactPosition(position)\n",
    "        elif to_res == \"?\":\n",
    "            to_res = UnknownPosition()\n",
    "        elif to_res.startswith(\"?\"):\n",
    "            position = int(to_res[1:])\n",
    "            to_res = UncertainPosition(position)\n",
    "        elif to_res.startswith(\">\"):\n",
    "            position = int(to_res[1:])\n",
    "            to_res = AfterPosition(position)\n",
    "        else:\n",
    "            position = int(to_res)\n",
    "            to_res = ExactPosition(position)\n",
    "        location = FeatureLocation(from_res, to_res, ref=isoform_id)\n",
    "        feature = FeatureTable(\n",
    "            location=location, type=name, id=None, qualifiers=qualifiers\n",
    "        )\n",
    "        record.features.append(feature)\n",
    "        return\n",
    "    # this line is a continuation of the previous feature\n",
    "    feature = record.features[-1]\n",
    "    if line[5:34] == \"                             \":  # old-style FT line\n",
    "        description = line[34:75].rstrip()\n",
    "        if description.startswith(\"/FTId=\"):\n",
    "            # store the FTId as the feature ID\n",
    "            feature.id = description[6:].rstrip(\".\")\n",
    "            return\n",
    "        # this line is a continuation of the description of the previous feature\n",
    "        old_description = feature.qualifiers[\"description\"]\n",
    "        if old_description.endswith(\"-\"):\n",
    "            description = \"%s%s\" % (old_description, description)\n",
    "        else:\n",
    "            description = \"%s %s\" % (old_description, description)\n",
    "\n",
    "        if feature.type in (\"VARSPLIC\", \"VAR_SEQ\"):  # special case\n",
    "            # Remove unwanted spaces in sequences.\n",
    "            # During line carryover, the sequences in VARSPLIC/VAR_SEQ can get\n",
    "            # mangled with unwanted spaces like:\n",
    "            # 'DISSTKLQALPSHGLESIQT -> PCRATGWSPFRRSSPC LPTH'\n",
    "            # We want to check for this case and correct it as it happens.\n",
    "            try:\n",
    "                first_seq, second_seq = description.split(\" -> \")\n",
    "            except ValueError:\n",
    "                pass\n",
    "            else:\n",
    "                extra_info = \"\"\n",
    "                # we might have more information at the end of the\n",
    "                # second sequence, which should be in parenthesis\n",
    "                extra_info_pos = second_seq.find(\" (\")\n",
    "                if extra_info_pos != -1:\n",
    "                    extra_info = second_seq[extra_info_pos:]\n",
    "                    second_seq = second_seq[:extra_info_pos]\n",
    "                # now clean spaces out of the first and second string\n",
    "                first_seq = first_seq.replace(\" \", \"\")\n",
    "                second_seq = second_seq.replace(\" \", \"\")\n",
    "                # reassemble the description\n",
    "                description = first_seq + \" -> \" + second_seq + extra_info\n",
    "        feature.qualifiers[\"description\"] = description\n",
    "    else:  # new-style FT line\n",
    "        value = line[21:].rstrip()\n",
    "        if value.startswith(\"/id=\"):\n",
    "            qualifier_type = \"id\"\n",
    "            value = value[4:]\n",
    "            assert value.startswith('\"')\n",
    "            assert value.endswith('\"')\n",
    "            feature.id = value[1:-1]\n",
    "            return\n",
    "        elif value.startswith(\"/evidence=\"):\n",
    "            value = value[10:]\n",
    "            assert value.startswith('\"')\n",
    "            if value.endswith('\"'):\n",
    "                value = value[1:-1]\n",
    "            else:  # continues on the next line\n",
    "                value = value[1:]\n",
    "            assert \"evidence\" not in feature.qualifiers\n",
    "            feature.qualifiers[\"evidence\"] = value\n",
    "            return\n",
    "        elif value.startswith(\"/note=\"):\n",
    "            value = value[6:]\n",
    "            assert value.startswith('\"')\n",
    "            if value.endswith('\"'):\n",
    "                value = value[1:-1]\n",
    "            else:  # continues on the next line\n",
    "                value = value[1:]\n",
    "            assert \"note\" not in feature.qualifiers\n",
    "            feature.qualifiers[\"note\"] = value\n",
    "            return\n",
    "        # this line is a continuation of the description of the previous feature\n",
    "        keys = list(feature.qualifiers.keys())\n",
    "        key = keys[-1]\n",
    "        description = value.rstrip('\"')\n",
    "        old_description = feature.qualifiers[key]\n",
    "        if key == \"evidence\" or old_description.endswith(\"-\"):\n",
    "            description = \"%s%s\" % (old_description, description)\n",
    "        else:\n",
    "            description = \"%s %s\" % (old_description, description)\n",
    "        if feature.type == \"VAR_SEQ\":  # see VARSPLIC above\n",
    "            try:\n",
    "                first_seq, second_seq = description.split(\" -> \")\n",
    "            except ValueError:\n",
    "                pass\n",
    "            else:\n",
    "                extra_info = \"\"\n",
    "                # we might have more information at the end of the\n",
    "                # second sequence, which should be in parenthesis\n",
    "                extra_info_pos = second_seq.find(\" (\")\n",
    "                if extra_info_pos != -1:\n",
    "                    extra_info = second_seq[extra_info_pos:]\n",
    "                    second_seq = second_seq[:extra_info_pos]\n",
    "                # now clean spaces out of the first and second string\n",
    "                first_seq = first_seq.replace(\" \", \"\")\n",
    "                second_seq = second_seq.replace(\" \", \"\")\n",
    "                # reassemble the description\n",
    "                description = first_seq + \" -> \" + second_seq + extra_info\n",
    "        feature.qualifiers[key] = description\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# scan function to avoid errors in ScanProsite\n",
    "from urllib.request import urlopen\n",
    "from urllib.parse import urlencode\n",
    "\n",
    "def scan(seq=\"\", mirror=\"https://prosite.expasy.org\", output=\"xml\", **keywords):\n",
    "    \n",
    "    #Borré la descripción de la función solo para que la visualización de los cambios fuera más sencilla\n",
    "    parameters = {\"seq\": seq, \"output\": output}\n",
    "    \n",
    "    ##Cambio para solo usar sig\n",
    "    if seq is None:\n",
    "        parameters = {\"output\": output}\n",
    "    ##\n",
    "    \n",
    "    for key, value in keywords.items():\n",
    "        if value is not None:\n",
    "            parameters[key] = value\n",
    "\n",
    "        \n",
    "    command = urlencode(parameters)\n",
    "    url = \"%s/cgi-bin/prosite/PSScan.cgi?%s\" % (mirror, command)\n",
    "    print(url)\n",
    "    handle = urlopen(url)\n",
    "    return handle"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='1'></a>\n",
    "## Ejercicio 1\n",
    "[Regresar al índice](#indice)\n",
    "\n",
    "Busca los patrones de Prosite PS00079 y PS00080 en la Ferroxidasa humana HEPHL1 e imprime la o las partes de la secuencia que entra en el patron y el rango (donde inica y termina el patrón en la secuenca)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "https://prosite.expasy.org/cgi-bin/prosite/PSScan.cgi?seq=HPHL1_HUMAN&sig=PS00079%0APS00080&output=xml\n",
      "\n",
      "Number of matches: 4\n",
      "\n",
      "Prosite    #\tSequence\t\tRange\n",
      "PS00079    1\tKWMITCQVSDHLQAGMLGQY\t341-361\n",
      "PS00079    2\tIFRVFCATMPHLSRGMGQIY\t694-714\n",
      "PS00079    3\tTWLLHCHVSDHIHAGMETTY\t1043-1063\n",
      "PS00080    1\tCHVSDHIHAGM\t1048-1059\n"
     ]
    }
   ],
   "source": [
    "from Bio.ExPASy import ScanProsite\n",
    "from Bio import SwissProt, ExPASy\n",
    "\n",
    "PROSITE_accession = \"PS00079\\nPS00080\" \n",
    "uniprot_id = \"HPHL1_HUMAN\"\n",
    "# Access ScanProsite file\n",
    "handle = scan(seq=uniprot_id, sig=PROSITE_accession)\n",
    "results = ScanProsite.read(handle)\n",
    "\n",
    "n = 1\n",
    "prosite = ''\n",
    "\n",
    "print (\"\\nNumber of matches:\", results.n_match)\n",
    "print(\"\\nProsite    #\\tSequence\\t\\tRange\")\n",
    "for result in results:\n",
    "    # Access protein sequence \n",
    "    handle = ExPASy.get_sprot_raw(str(result[\"sequence_ac\"]))\n",
    "    record = read(handle)\n",
    "    # Counter to identify match number per accession ID\n",
    "    n = n+1 if prosite == result['signature_ac'] else 1\n",
    "    # Save sequence\n",
    "    prosite = result['signature_ac']\n",
    "    sequence = record.sequence\n",
    "     \n",
    "    print(\"{}    {}\\t{}\\t{}-{}\".format(result[\"signature_ac\"], n, sequence[result['start']:result['stop']], result[\"start\"], result[\"stop\"]))\n",
    "    \n",
    "    \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='2'></a>\n",
    "## Ejercicio 2\n",
    "[Regresar al índice](#indice)\n",
    "\n",
    "Obtén el nombre de las enzimas de una lista que contenga accesion numbers de SwissProt (los que quieras) utilizando expresiones regulares"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enter accession numbers separated by a space:O23729 O46647\n",
      "\n",
      "\tEC number\n",
      "O23729\t2.3.1.74\n",
      "O46647\t3.1.1.34\n"
     ]
    }
   ],
   "source": [
    "from Bio import SwissProt, ExPASy\n",
    "import re\n",
    "\n",
    "accession_num = input(\"Enter accession numbers separated by a space:\")\n",
    "accession_num = accession_num.split()\n",
    "print(\"\\n\\tEC number\")\n",
    "# Search EC number for each accession number entered\n",
    "for accession in accession_num:\n",
    "    handle = ExPASy.get_sprot_raw(accession)\n",
    "    record = read(handle)\n",
    "    # Extract EC number from record's description\n",
    "    EC_num = re.findall(\"\\d+\\.\\d+\\.\\d+\\.\\d+\", record.description)\n",
    "    print(accession + \"\\t\" + EC_num[0])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='3'></a>\n",
    "## Ejercicio 3\n",
    "[Regresar al índice](#indice)\n",
    "\n",
    "\n",
    "Escojan un organismo de los presentes en Cellosaurus y después escojan una categoría para graficar los porcentajes.\n",
    "\n",
    "Por ejemplo, en la categoría de enfermedades cuáles son las más representadas o en sexo cuáles son los porcentajes (x% mixed sex, y% female, z% male)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sex ratios for Homo sapiens cells on cellasaurus:\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAWQAAADvCAYAAADFG66PAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMi4yLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvhp/UCwAAIABJREFUeJzt3Xd81dX9x/HXuTc3kxD2CDuAl6UStiiu4ohbobbuulpH1fqrv/5QO0xrNW3tUFurllZtrdZRRZsawEUQUEAJMpJcAmHvANnrjvP743vBIATIzb33fO/N5/l43AfJzf3e+7kxvnNyptJaI4QQwjyH6QKEEEJYJJCFEMImJJCFEMImJJCFEMImJJCFEMImJJCFEMImJJCFaEEpNVgpteZr9z2ilHrAUD2ZSqk3W3z+qlJqlVLqfqXUz5VS09vwXEe8N2EvCaYLEEK0Tmu9A5gJoJTqA0zUWg8zW5WIFGkhC3GClFILlFK/UkotU0qtU0pNC94/OnjfymDrdXiwNVqqlPqnUqpEKfWmUio1+PjxSqlCpdQXSql5Sqm+wfuHKaU+UEp9qZRaoZQa+rVW7XygX/B1pimlXlRKzTzOc44PPt+XwN1R/6aJNpFAFqJtErTWk4AfAD8L3ncH8KTWeiwwAdgWvN8NPKO1HglUA3cppVzA08BMrfV44G/AL4OP/yfwJ631qcBUYOfXXvsyYIPWeqzW+pODdx7nOV8A7gk+p7A56bIQ4nCt7SVw8P63gv9+AQwOfvwp8LBSqj/wlta6TCkFsFVrvTj4mJeBe4G5wBjg/eBjnMBOpVQ60E9r/TaA1roRIPiY43G38pxdgC5a64XBx/0DyDmRJxRmSCALcbh9QNev3dcN2Bj8uCn4r5/g/z9a61eUUkuBi4H3lFLfA8o5Mtw1oIC1WuvTWn4hGMihau05u7TjOYUB0mUhRAta61qs1uW5AEqpbsCFwKLWrlFKZQHlWuungHeAU4JfGqiUOhiS1wafwwP0PHi/UsqllBqtta4Btimlrgjen3Swz/kEtPaclUClUuqM4OOuO8HnE4ZIIAtxpBuBnyilVgIfAbla6w3HePzVwJrg48cAfw/e7wHuVkqVYLW6/6y1bsaaNfGr4EDbSqz+YoAbgHuVUquAJUCfEyn2OM95M/CnYG0n1P8hzFGy/aYQ4aeUGgzka63HGC5FxBBpIQshhE1IC1kIIWxCWshCCGETEshCCGETMg9ZxJSq3Fwn0APoCaRg/QwnAM7OM3+hldOvAV/wVoU1r/gAI7TfUMlCnDAJZGEbVbm5DmAYcDJwEta0r17BW+/gv91p7S87h78RSD7KVzSlqhLYjxXQFVjLm8uDtzJgHSN0fRjfjhBtJoEsjKjKze2LFbwtbyOxWr0hCQRIcDqP+iWFNQ+4KzC0lcs1pWoLUIK1LPpT4DNG6H2h1iNEW0kgi6gIBvD04O0bQL9wv4bW7Vr4oIBBwduFh+4tVWVY4XzwtooRMjVJRIYEsoiIqtzczsDZwHSt9TeUUqMi/qKq1Y2B2mN48HZj8PO9lKp5QIHHf0WBe/TbByLwmqKDknnIImyqcnO7YC0jvlZrfbpSKqq/8NNmPuJLcEavkfFsY8nmSj20HHgbeHtWtmvb8a4R4lgkkEW7VOXmJgKXaK2vBy5WSiWaqiWagVzvTd30lK9ycIu7NPA51jabL83KdlVFow4RX6TLQrRZVW6uAqZpra/XcLVDqYwT3Lc3bny665xkehx2lwImBm+P5xV5XwGemZXtKjJQnohREsjihFXl5iYDN/kDgf91OhxDlVIddvuw8qTvHGtRVSpwG3BbXpF3KfAM8NqsbFfTMa4RQrosxPFV5eZ29fr99yql7k1wOLqZrqc10eqyqGtKaHjKX52sVEJbfh9VYB2n9OysbFd5hEoTMU4CWbSqKjd3ULPf/yOnUrc4HY6jLbiwlWgF8so92Vvnpi8dEOLlGpgDPDwr21USxrJEHJAuC3GEqtxcd5PP96jL6bwy0dnKUosOrMT/rfbOd74SuCyvyPsi8IjMzhAHSQtZHFKVm9u1pqnpN2mJid9xKBVzQRyNFrLPrwK/r9vp87u6hWs2SSPWidGPz8p2yZzmDk4CWVCVm5uwv77+R52Skh5KdDrTTNcTqmgEcvn+/jtfTynvG4GnrgTygKdmZbsaIvD8IgbI9psd3KZZs66sb27e1C019ZexHMbRUtJ4cXOEnroLViCX5RV5b88r8sbcXyii/SSQO6idDz88euuDD37WNSXlrdTExLDvKxGvNiTf3jXCL9EPeB5YlFfkdUf4tYTNyKBeBzPv1lsdAzIyHu+Tnv7DVIdDWmFtsLcm/UB9yimRDuSDpgBFeUXeHwN/mJXtCkTpdYVB0kLuQN6/7bZThnbv7umXkfEjp4Rxm5XWnh7t5dApwG+BBXlF3ta2DRVxRAK5A8hxu9XS738/d2xm5uc909KGma4nVq1z3pRq6KWnAV/mFXnvzivydtTFkR2CzLKIc/Nvu23IwIyMt/t27nyq6VoiLZKzLOqbEhqf8lUl4XCZDsSPgFtmZbs2G65DRIC0kOPYojvuuOvUPn3WdoQwjrSyqtF7bBDGAOcCq/OKvDeYLkSEnwzqxaHrsrOTv3/aaf8c06fPVR1tF7ZIKfVfbafGSzrw97wi7wTgh7OyXT7TBYnwsNMPmQiDvJwc9/+dddbqUb17SxiHid9PYEvazT1N13EU9wLv5xV5exz3kSImSCDHkeeuuurSb5966tJBXbvKwF0YbanK3ON39UgyXUcrzga+yCvyjjVdiGg/CeQ4kON2O1785jcfvHzUqDe6pqRkmK4n3pQ05TSaruE4BgKf5BV5LzJdiGgfCeQYl+N2J982ceILl40a9WiKy2XXVlxMW5/03WgtBmmPTsC7eUXeu0wXIkIng3oxLMft7nLXaae9duaQIeebriVe7atNq6xPye5iuo4T5AT+FFxE8sCsbJfMaY0x0kKOUTlud/87p0zJlzCOrJLaqbF4WOn/AM/JIpLYIy3kGJTjdg+8a8qUN8/MyppoupZ453HcFKvdQLcDXuBu04WIEyct5BiT43YPvvu0096SMI68hmZn0960K3ubrqMd7sor8v7BdBHixEkgx5Act3vIPVOnvj1tyJDxpmvpCMoqR9lldV573JdX5P2N6SLEiZFAjhE5bvfQH5xxxpzTBw+W+aZRUur/pukSwuWBvCLvL00XIY5PAjkG5Ljdw+6YMuX1KQMHnmK6lo7CH7Dt6rxQPZRX5H3EdBHi2CSQbS7H7R5+xejRfzk7K2uc6Vo6kq1Vfff4XL2TTdcRZj/LK/I+bLoI0ToJZBvLcbszTxs06DffPPnkaaZr6WhKGi9oMl1DhDwqO8XZlwSyTeW43Rkn9ejx8+9NmnSBnO4RfRuSbu9suoYIej6vyCsDwzYkgWxDOW53Uu9OnR64f9q0mckuV7z92Rx2/kCAac8+y2V3cdRfXK8XwKhLYPQlcO0D1n2ejTB+BpxyOXxaZN3n88H0m2FbRUpVbcrEWFguHapk4O28Im889ZHHBQlkm8lxux1JCQm3PHDmmbfJRkEn5s+ffYa7x9F3oCzbBI//BRb/E9bmwx8etO5/7jV48iF47zl44oXg8/wLrr8MNnunVkancqMGAK/nFXllcZiNSCDbz2X3n3HG/wzo0qWP6UJiwfaqKuaXlXHDuKOPef7lDbj7Guga/NXWq7v1rysB6huhvsH6uLIa/vMx3Hg5rHPeGKur89rqbOAJ00WIr0gg20iO233aZSNH/nBsZqbsZ3yCHpw7l5+fdx6OVjbjX7cZ1m2C06+FKd+CuZ9Y9999LTz2HNz0IDz0PfjFn61/m33Opt1pM3pF7x0Yd58M8tmHBLJN5LjdWYO7dv3BzFNOkSXRJ2iux0PPtDTGZma2+hifD8o2w4KX4NXfwu0/tVrDAzNhwd/h039BajJs2wUjs+Dy+9L8rzz4Hcfezeui+E6Mk0E+m5BAtoEctzvVqdSd951++rREp7Oj/Lncbku3bqXA4+Hk3/+eW998kwXLUNf/6PDH9O8Dl50LLhcM6Q8nDbYCuqWH/wCP3gdPvQyTLrq0Nue+x/nw+Uej9j5s4OAgnxwFZZgEsmE5brcCvn37pEnf6Nu5c1/T9cSSn02fTvEPf8jq++/nrzNncvYk9Mu/PvwxV3wDFiyzPq44YHVfZPX/6uuFyyCzFwwfDHX1sC/9sgzlcOBtrI/W27CLAcBTpovo6CSQzRs/sX//b56ZlSV7VITJT5+Cdz+yPr7gDOjexZr2ds5N8JsHoHtwQpvW8Oiz8JM7rc8vOr97xTtPPp700r2XM+2G/zFTvFnX5BV5LzZdREemtJZDBUzJcbt7ZCQn//qJiy+ekZ6UFM8LEaIibeYjvgRn6Ht8z91x46aVXWcPDmNJsWgbMGpWtqvGdCEdkbSQDclxuxOA2+6ZOvV0CWN72JB0q8z7hv7Ar0wX0VFJIJtzwZSBA88e06fPSaYLEXCgLqWqJvW0eF6d1xZ35BV5Zf8UAySQDchxuwc4lfrmjePGZZuuRVhKayZ3hNV5J0oBs/OKvLJsP8okkKMsx+12ANddm509tFtqakdagGBrHnVjoukabOYk4Kemi+hoJJCjb3yPtLSx5w0bNsF0IcLS6HU07+o0M5bPzouU/80r8srsnyiSQI6iHLc7FbjhzilTRiUmJMifgzaxofKkPTiS5f+FIyUAfzJdREciP4TRlTOuX7/Bo3r1Otl0IeIrJb4ZMvezdVPziryXmS6io5BAjpIct7uvgotvnjBhvGplIxwRfYEAenPabbJk+NgeyyvySlZEgXyToyC4PPqaC9zuzJ5paa3vhCOibnt1zz1eV78U03XY3GhAdoSLAgnk6DgZOPWSESOkq8JmSurPbzRdQ4zIzSvyukwXEe8kkCMsOM1txvRhwzr3kNax7axPujXddA0xYhDwHdNFxDsJ5MgbCQy6ZOTIox9pIYyprE+urk49o5vpOmLIg3LkU2RJIEdQsO/4qjMGD07vk54+0HQ94nCl1ZMOmK4hxgwBrjNdRDyTQI6sk4BhV4wefarpQsSRPOp6WZ3Xdg/lFXmPerq3aD8J5AgJto4vH9+vX0r/jIyhpusRh2vyOrw7O31Llq633UnARaaLiFcSyJEzBBg54+STR5suRBypvGrYbhwp0tILzW2mC4hXEsiRc2mvtDQGd+kywnQh4kgl3isDpmuIYRflFXnluLEIkECOgBy3uy8w9vLRo/s6HA75HttMIIDemPpdWZ0XugRkClxESFhExmTAPy4zU3bKsqGd1T32ehMHpJquI8bdklfklT0AwkwCOcyCRzNNn9i/v7Nraqps6WhDxQ3TO9yR0hEwDDjLdBHxRgI5/EYAaecPHy6DeTa13nWLnGEYHjK4F2YSyOF3VlJCQtNJPXvKvhU2VFWfVFOVdraszguPGXlF3i6mi4gnEshhlON2ZwDjLx05skeSbEBvS57qCftN1xBHkpGVe2ElgRxe2QCTBgyQ1rFNlarrZMey8JJtOcNIAjlMgivzzktJSKjO7Nx5iOl6xJGafQ7vzrRrZaA1vCbmFXm7my4iXkggh08m0PesrKzuCQ6HtMJsqLwya492psrqvPByANNNFxEvJJDDxw0wNjNzmOlCxNGVeK/0m64hTl1guoB4IYEcPpOAmiHdug03XYg4ktawMeV2WZ0XGeebLiBeSCCHQY7bnQoMH96jh8pITpb+NBvaWd19b3PSYFmdFxn98oq8Mu8+DGwZyEopv1JqZYvb4Ai+1neUUn9s59MMBdQZgwdnhaMmEX4l9efWma4hzkm3RRjY9TiWBq11LO0DcQrgG9Gzp3RX2FSZ65ZOpmuIcxcAvzNdRKyzZQv5aJRSTqXUb5RSy5VSq5RS3wvef7ZSqlAp9Y5SqlwplaeUuk4ptUwptVopNTT4uEuVUkuVUkVKqQ+UUkdMf1JK9VRK/Tv4GsuVUqcfr67gdLeJiU5npUx3s6fqhqTayrRvSP9xZJ2ZV+RNMV1ErLNrIKe06K54O3jfrUCV1noiMBG4XSl1MABPBe7AOlD0BuAkrfUkYDZwT/Axi4ApWuts4F/Aj47yuk8Cvw++xozg9cfTF+g8NjMzw+V0ynQ3G/JUj9tnuoYOIBk403QRsS6WuizOB05RSs0Mfp4BDAeageVa650ASqkNwPzgY1YD5wQ/7g+8ppTqCyQCG4/yutOBUUod2lWws1Kqk9a69hi1DgMY1atX5om+ORFdpVwrvyijYzIwz3QRscyugXw0CrhHa33Yf3Cl1NlAU4u7Ai0+D/DVe3wa+J3W+t3gNY8c5TUcWK3oxjbUNRJoGNS1a782XCOipNmnfDvSrpOz86JjjOkCYp1duyyOZh5wp1LKBaCUOkkpldaG6zOA7cGPb2rlMfP5qosDpdSJDCyeBNT06dRJWsg2tLFyyG7t7BRLDY9YJoHcTrEUyLOBYmCFUmoN8Bxta+E/AryhlPoCqGjlMfcCE4KDhsVY/dKtynG704GuSQkJ3oyUlJ5tqEVESYn3clmdFz3D84q8iaaLiGW2bDlorY+YoqS1DgAPBW8tLQjeDj7u7BYfH/qa1vod4J2jPO+LwIvBjyuAb7Wh1H6APrl37x4OpWLpl1uHoDWUp3xXFupETwLWFgKrTRcSq2wZyDGkD+AY1qNH1HYQ8wcCnP3882Smp/Paddfx/XfeoWjHDrTWDOvenWeuuIJOSUmHXdPs8/GD/HxW7tiBUoq8Cy9k2pAhNPl8XPvqq+yorubWiRO5bdIkAO57911unjCBsZmx3Quzq6ZrRXPSUJnuFl1jkEAOmbTq2mco0DCgS5eoBfKfP/sMd4+vMuaxCy5g8Z13suSuu+ifkcFfli074pqXVqwAYMlddzHnhhv48fz5BAIBPly/nikDB7L4zjt5bdUqAFbv2oVf65gPY4CSunOONTtGRIb0I7eDBHL7ZAF13VJSukbjxbZXVTG/rIwbxo07dF/nZOtgEq01jT4fLabsHeLZu5czh1hTtnt26kRGcjJFO3bgcjpp8HrxBgJorQH45Ucf8fA55xzxHLFIVucZIYHcDhLIIcpxu11YXRb1nRITo3Jo5oNz5/Lz887D8bXQvWvOHE564gnWVVTw3WC3Q0tjevemwOPB5/ez6cABVu7Ywbbqas7JymJLZSXTZ8/me5Mn815pKaf27UvfzrF/BmhNQ2Ld/tTp0l0RfRLI7SB9yKE7eLijTk1MTI/0i831eOiZlsbYzEw+2Xj4mpZnrrgCfyDAj957j7fWruX67OzDvn59djaeigrOfv55BnTpwuQBA3AqRYLTyeyZ1jobr9/PVf/4B69ccw0PzZ3Ltqoqvn3qqVw0YkSk31pErKvO3qc6O9oyLVKEx5C8Im/qrGxXvelCYpEEcujSAe1QSqW4XBH/03jp1q0UeDzMLyujyeejpqmJ7/773zw/YwYAToeDq8aM4anFi48I5ASnk8cvvPDQ5+fPns2w7odPPpi9fDnfPvVUPt+2jc7Jybxw/vlc+tJLMRvIpfoaORnEDIW1ncAG04XEIgnk0KUDKrNz57RoTHn72fTp/Gy6dVLOJxs38sclS3juqqso37ePrO7d0VpT4PEwvMeRf6XXNzejgbTERD7esAGnw8GIXl8tXqtsaGDeunW8df31FKxbh0MpFNDo9Ub6bUWE16d82zvdIGfnmdMLCeSQSCCHLh1wZHbuHPHuitZorblzzhxqmprQWjOmTx9+e/HFALxXWkrRjh08fO657K2rY8bLL+NQir7p6Tx31VWHPc+vCgv54bRpOBwOvjF0KLOXLWPqmjXcPGGCibcVOo0C2FQ1aHcgNV2WspsjS9VDpA6Orou2yXG7rwAuuXTkyNTrsrOvMV2PgNQZjwRcCTje3XnvluIuTww0XU8H9t1Z2a6/mC4iFsksi9D1BJq6p6YaayGLw+lgC7k85faoTEMUrZJtBEIkgRy6HkBTRnKyzHW1kV3VXfY1Jrnll6RZGaYLiFUSyKHrATS7nE7ph7cNpUvrzpLVeebJL8QQSSCHrjPQLJsK2YdS4Em4RU6WNk8COUQSJiEInqPnAgISyPZR15RUuz/tAum/NC/2l3oaImESGgVoAKdSsgDBJtZVj9svvx9tQcZVQiQ/vaE5FMIOh0O+hzZRynXSn28PPtMFxCoJk9Ac+r5Jl4U91Cmnd0ena2R1nj205UxK0YKESWgcBLssJJDtobDP6B0kdJLTpe1BAjlEEiahkRayzWwYNklmV9hHg+kCYpWESWgOtZD9gYAcomlYcXL3yoSuvWR2hX1ICzlEEsihOTSo1+D1SmvAsE+HTq4yXYM4jARyiCSQQ+PDmvpGvQSyUbU4fQ0DR/c1XYc4jPw/ESIJ5NA0Evze1TY3yw+fQYV9R+90ulyJpusQh5EWcogkkENQ4PH4gCbAWdvcLD98BpUPm5RsugZxBPl/IkQSyKGrBRJqGhulhWxISXK3AzKYZ0tynl6IJJBDVwu4KiWQjVkydEq16RrEUW0xXUCskkAOXQ2QcKChQQLZAGswb1Qf03WIoyo3XUCskkAOXTXg2l1TU2e6kI6osM+oHU5XYpLpOsRRSSCHSAI5dFVAQmVjY3OD1yuhHGXlwyenmK5BHNXuWdku+f8hRBLIoTuAtScyVY2NFYZr6VBkMM/WpHXcDhLIodt78IN99fUSyFH0aZYM5tmYBHI7SCCHroLgfha7a2okkKOkFqevflBsD+b96uLh/OHqbJ769gT+eN0UAFa//ya/n3kqD41PYlvxF8e8PuD389Q1E3nx3isO3fevh2/kyavHMe/pHx+676PZj7H243ci8yZaJ4HcDrKhd+j2EfyFtrWqSgI5Shb2GbXT6UocYLqO9rr9ufdJ69rj0Oe9h47m+ide5+1f3n3caxe/+jS9hoygsbYGgJ3rVuFKSuG+11fw1ztzaKypormxnq2rl3HubQ9F7D20YkO0XzCeSAs5RAUeTz3WTIvE9fv2SSBHyYY4XZnXK2skPQe7j/u4qt3b8HxSwMQrbjl0nzPBhbepgUAggN/nRTmdfPBsLtPv+FkkS26NtJDbQQK5fbYCqesrKip9gYAcWxNhpcndDiR06x3zg3lKKf5290U8fe1klv17dpuuzX/ih+Tc9ziqxclhvbJGktalJ3+8dhIjz7yYfVvXowMB+o3MDnfpJ2KdiReNF9Jl0T6bgZM0VFY3NlZ0S02N6b5Nu1tiDeZ1NV1He33vbx+T0asftfv38Nc7c+g52M2Q8dOOe13Jwv+S1q0X/UaNo/zzwsO+dun//vbQxy/ddwVX/vgZPp79ODvLVjFs8nQmXXVr2N/HUZTPynbtjsYLxStpIbfPNoK/1HbW1Gw3XEtcq8fhj/XBvIMyevUDoFO3Xow+53K2rl1+Qtdt/nIJJYX5/Ori4bz64PWUf/4xrz1802GPKV7wLv1GjqOpvpZ928q59levsuaDt2huiMr2Ep9E40XimQRy+xyaabFu797NhmuJawviZGVec0MdTXU1hz4u++wDeg8dfULXXnjPL3lw7kb+779lXPP4y2RNOIdv/fKlQ1/3e70sfuVpzrzpAXxNDSilAAgE/Ph9zeF/M0eSQG4nCeT22Ym1Ub1aunXrJsO1xLUNwybHfBgD1O7bzbO3nM2T3xrPn26YyogzcnCffgFrP5rD4xcOYcuqz3jp3sv5210XA1C9dwcv3HPZCT33p6//mXGX3EBiSip9hp9Cc2M9f7g6m34js0lJ7xLBd3XIwmi8SDxTWmvTNcS0HLf7F0AKUPuXGTPuTU9Kivk+TrspTe524OPz75Dvq41prXc/OC4xLrqUTJIWcvutBDIAtlVVbTJbSnxakjVZVubZnFJKuivCQAK5/dYR/D6WVVRIP3KYWYN5o6XlZX8SyGEggdx+mwh+H5dv2yaBHGYLesfHYF4HIIEcBhLI7VTg8dRgDe6llVVUVNY1N8uR9GG0YfjkuFyZF0+01tXAl6briAcSyOGxEugCsOnAgfWGa4kbnuSulfGwMi/eKaU+mpXtCpiuIx5IIIeHh+D38tPNm9cYriVuLB4yWf7aiA1vmi4gXkggh8fBDVUcH23YsLnB6601Wk0cqMfhrx88RgbzbE5r3QT8x3Qd8UICOQyC/chrgG4BrXVZRUWx6ZpiXaEM5sWK+bOyXSc0LVEppZVSL7f4PEEptVcplX+c684+3mPihQRy+CwCUgEWb9681nAtMW/98EkSxjFAKfVGGx5eB4xRSh08D/E8QPaAaUECOXyKsfa1cCwsL99S39xcY7qgWGUN5vXpZboOcWxa6wZgThsvew+4OPjxNcCrB7+glJqklPpUKVWklFqilDpig2ilVJpS6m9KqWXBx10e8huwIQnkMCnweOqwpv5018C6igppJYdoSZYM5sWIt2dlu9ra8PgX8G2lVDJwCrC0xddKgWla62zgp8BjR7n+YeAjrfUk4BzgN0qptLaXbk8SyOG1mGC3xaJNmySQQ1CPw183SAbzYoFS6u9tvUZrvQoYjNU6fu9rX84A3lBKrQF+DxxtG7zzgVlKqZXAAiAZGNjWOuxKAjm8SgA/4Fy0adO2yoYGOdqpjWQwLzYEAoFdwAchXv4u8AQtuiuCfgF8rLUeA1yKFbZfp4AZWuuxwdtArXVJiHXYjgRyGAXP2SsCugMs3bp16bGvEF8ng3mxQSn191nZLn+Il/8NyNVar/7a/Rl8Ncj3nVaunQfco4KbPSuljJxTFSkSyOG3kGC3xRurVn3Z5PM1GK4nZshgXmzQWvuUUk+34/ptWuunjvKlXwOPK6WKaP14uV8ALmCVUmpt8PO4Ifshh1mO2+3E+sEKAHUPnHnm9An9+59uuKyY8MLICzc3Dh83yHQd4th83uZ//nhS2vWm64hH0kIOswKPx4+1cqkHwL9Xr14W0FrW+R9HPQ5/3WDZZtPutNY6wZX4qOk64pUEcmQsA7yAa+OBA9Xl+/fLyr3jsAbzkqT/2Ob8zU3zZmW7Sk3XEa8kkCMgOLg3H+gN8F5p6WdmK7K/9cNkMC8WJCQl55quIZ5JIEfOQsAJOJZs3rx9T23tNtMF2dW6pC6QNxNKAAANy0lEQVRVCd1lMM/ufM2Ny2Zlu6RxEUESyBFS4PHsAT4HegHMW7dOTuRtxeKsKZWmaxDH50xIfMR0DfFOAjmy5hOc3P7f0tKy3TU1WwzXYzsymBcbfM1NngfHJxWYriPeSSBH1nqgjOCMi7fWrv3QbDn2s7D3yJ0ymGd/DmfCz03X0BFIIEdQgcejgTeAdEAVlpdv2VJZWWa4LFspGzYp0XQN4tiaG+u/dDidX1/mLCJAAjnyyrDO3OsF8M+iovcDshoHODiY11cG82xM64AmELh1VrZLfmajQAI5woKt5H9j9SU7vty5c2/x7t1fGC7LFhZnTZbBPJurr9z3xk9P7yo/r1EigRwFBR7PFqwTRfoCzF6+/ONmv7/JbFVmWYN5ss2mnfmam+oSU9LuMF1HRyKBHD0HT1ZI3FVTU79o06YFJosxTQbz7K+hpvKRn5yWccB0HR2JBHKUFHg8+4B3gD4Af12+fGlHXiwig3n21lhbvT69e+/fma6jo5FAjq4PgWog3R8I6L8sW/aOLxDwmS4q2spkMM/WtNb4mptunpXtkk2xokwCOYoKPJ4G4K9Y85Idq3ftqljSAbsuFg2RwTw7q6/c9/aj38hcZLqOjkgCOfrWAIVAP4Dnli1bsre2tsMchR4czOttug5xdM2N9VWJKWm3ma6jo5JAjrLgNLjXgVqCXRfPL1s2p6N0XSzsNWKHMzHpaGelCcN0IKD3bd3wnZ+c1nm/6Vo6KglkAwo8nlpgNi27LjZvXmC2qugoGz5ZBvNsatf6Na8+efW4Ocd/pIgUCWRz1mBt0ZkJ8NzSpUvifdZFcDBPuitsqGrP9g3FH79zs+k6OjoJZENadF3UE+y6+N0nn7ze4PXWGS4tYhbLYJ4tNTfUNWwv/uLS95/9ebPpWjo6CWSDCjyeGr7qunBuOnCg5qUvvng9EAjE3XSjBlSgVgbzbEcHAnp7SdF9f79/RonpWoQEsh2sBvKBgQALysu3fLhhwzyzJYXfwl4jt8tgnv3sKS+Z89yt5/zFdB3CIoFsWLDr4m2sYM4E+Ovy5ctK9+xZabSwMFs3fLIsk7aZmopdm0sXvXet6TrEVySQbaDA4/EBzwM1QDeAXxcW/reirm6H0cLCpCwxQ1bm2UxTXU3d1jXLLit48qFG07WIr0gg20SBx1MNPAWkASn1Xq/vtwsXvtYYB4N8ss2mvXibGptLFubf+vf7Z6wyXYs4nASyjRR4PJuxWsp9AefGAweq//b556/5AgGv4dJCZg3mnSyDeTYR8PsCaz5869er5r3+uulaxJEkkO1nGfBfgoN8Czdu3PryihWv+QMBv9myQrOw10hZmWcTWmtKP3nvH6vnv5FbXJgvJ4DYkASyzbQ4YWQVwVCeu27dhtdXrXojoHXMTYcrGzbJZboGYSn79P3/fj7nhTuLC/M7xDL9WCSBbEPBQb5nsE6t7g/wTnGx5521a+foGDqPrywxo8rZI1O6K2xgw7KPP/ns9T9fV1yY32C6FtE6CWSbCm7V+RSwneDRT6+tWrV6rseTb7SwNpDBPHvYvHLJisWvPHV1cWF+lelaxLFJINtYcBOi3wH7CJ408tKKFSs+ioGFIzKYZw9b1yxfU/jib64oLszfZboWcXwSyDZX4PFUAU9gnTTSC+D5pUs/W7hx4wdGCzuOT3rKNpumlX32wecfz37s8uLC/K2maxEnRgI5BhR4PPuB3wBNWPte8Mynny5+t7h4jl33vVg3fLIM5hmiAwH95bzXFn36rz9dX1yYX266HnHiJJBjRIHHsxf4NRAg2FJ+ZeXKL/++YsUrXr/fVrt0rZfBPGP8Pq9v2Vuz535Z8K9bigvzPabrEW0jgRxDCjyencCjWN0XmWBNiXtq8eIXGrzeWqPFtbBIBvOM8DY1NC56+Q9vehYV3F5cmF9muh7RdhLIMabA49kDPAZswpqnrJZv27brsY8//mtVY2OF0eKQwTxTGmuraj+e/fgLm1cuubu4ML/DnNEYbySQY1Bw34vfASuAwYCjrKKi8ifz5v1td22t0QEcGcyLvpp9u/e//+fcJ3eVrf5RcWG+nIcXwySQY1SBx9MIPAvMxwpl1566uob/e++9l1bv2rXcVF0ymBddu9av2TTv6R//4sD2jb8oLsy3TbeVCI2KoYVf4ihy3G4F5ADfBnYCDQAzxowZdfno0ZclOp1R24d4Q1JG9fwL7u4crdfryPw+n3f1+28sXzXv9T8BrxUX5sfkXificBLIcSLH7Z4IfBdoBvYCuHv27HrP1Kkze6SlZUajhhfd521qcE8cHI3X6sjqDuytWPjS7xbt3VT6HDBPNgqKHxLIcSTH7e4L3IW1/8VWIJCUkOC8/4wzzhubmTk5kq/diArMvvB/mqX/OLK2rPqseNHLT77va278Y3Fh/nrT9YjwkkCOMzludxJwNXAesAvrVGsuHzXKfdWYMVckJSREJDDf7zli2/rTruofiecW4G1qbPjinRc+W7dk/hvAP6S/OD5JIMehYL/yeOB2wA/sARjctWv69yZPzhnSrdvIcL/mH0+7abezZz+Z7hYBlbu2bi984dcLq3Zv+zOwSLoo4pcEchzLcbt7A3dgzcLYDvgALhkxYvgVo0df1CkpqUs4XkcG8yLD29RYX1KY/+WXc1/9UAcCz8r84vgngRznctzuROAy4GKs7ou9AJ0SE113TJly5rjMzKkOh6Nd0x9fcp+3ud49cVD7qxUAOhAIbC9Z8eWnrz2zrqH6wLvAv4sL85tM1yUiTwK5g8hxuwcCNwHDsPqWGwDG9u3b8zsTJlzSJz19YCjPaw3m3d/kTExOCV+1HVfV7m2bPnvj2ZW716/dBLwIrJIuio5DArkDyXG7ncBU4FogAWveckAB12Vnjz136NBzUxMT09vynDKYFx5N9TWVq+a98XlJ4X82A3OAD4oL8xtN1yWiSwK5A8pxu7sAM4FpwAGgEiA5IcF5fXb2+NMHD56W4nJ1OpHn+uPUm3Y7e8hgXqj8Pp9304pPvlj67+fX+5oaPwXeKC7M32u6LmGGBHIHFZyJ4QZuBnpj9S3XAaS6XAnXZWePnzpo0BnHCubyxM7V8y78vgzmhcDv83p3lK5c+cW7L26s3rNjPfAPwCPdEx2bBHIHl+N2u4CJwDeBrrQI5rTExITrsrMnnDZw4BkpLlfa1699yT19c717kgzmtYGvualxW/Hnyz+f8+KW+sqKauB1rKlschK0kEAWluBsjEnADKALVjDXgzUj49qxY8dP6N9/Yufk5G4gg3lt1VRXc2Dzl0u+KPrvKzub6qqbgA+AguLC/GrTtQn7kEAWhwkG8xSsYO5Mi2BWwDdOco8ZO3jIBftHTj2w4bSrBpirNDZU792xad2S+StKFvynSutAM1AALJAToMXRSCCLowouwZ4CXIUVzLXA/trkTuOr0rqldskauXLUOVdm93WfOjYpLT0sC0ziRVNdTeWe8pK1pYsKNu70rPRhfe/+C3xSXJhfZ7g8YWMSyOKYgn3Mo4ELNbgrMnqPb0hMXa6dCdZG+EoxfMp5g4dOOmds9wFDRzoTXIlGCzakuaGuek95ydoNyxcUb1652AckAVuAfODL4sJ8W517KOxJAlmcsBy3u9+eLn1nNiR3cgMpWIN/+wANkJCY7MyaeNbA/qMmDO8+YOiwlM5de5qsN9KaG+pr9m7yFJd/vqB44xefNIBOwdo75HPgQ2CDzJoQbSGBLNps1FmXJAKjgHOAk7G6l2ux5jMf2ii9x6DhGVkTzh7eK2vUsIze/bKcCa6YPk0k4Pf56g5U7KjcuWXr1jXL1m9Y9nGt1oE0rPe8ElgMlBYX5jeYrVTEKglk0S6jzrqkG1Y4Twr+q4AAsJ/g8mwAV1KKM2vi2YP6jRw3rHOvfpmpGd16J9h872RvY0NdTcXOrfu3b9yyc92qrVtXL93la27KADphvcdVwCKgpLgwv95osSIuSCCLsBl11iXJwBCsVvNkrHnN8FXrOdDy8d36Z3XuM2xM7279hvRO75XZu1PXnr2TO2V0V+3c7KitdCCgmxvrq5vqairrKysqKras37pt7edb9pQX1wIZQCpWt4wGSoCFWCEsexKLsJJAFhEx6qxLFNALOAmr9XxwD+aDLei64M3b8jpXUoqzr/vUnt36D+2Z0rlLWlJqp9TElLRUV0paqispJdWVlJyakJicmpCYlHIiwR0I+P0Bn8/r93mbmupqKhtrqyrrq/ZX1h3YW1m9Z0flgZ2bKvdv21gd8HkDQBrWjBIXVvg2YgXwaqwBuu0yOCciSQJZRMWosy5Jwgro3sBAYDjWPs1JWOHnwOriqAeasPZuPuYPZ1qXHskpGV2TALTW6EBA60BA+33egLehzttUX+sN+H0tn8MBJLe4ubB+ORx8/T1Y4evBCuC9MignokkCWRgTbEV3wQrpPsBQrJDugtVNoPiqq4Dg547g519faqxafF21uObg9Qfv2wvsxtrpbidWV0olsF+6IIRpEsjCloJhnYwVzF+/pWN1LYDVwj3YyvVjdTM0YXWFNAdvB0O3trgw/7B+bCHsRAJZCCFsIqqj2UIIIVongSyEEDYhgSyEEDYhgSyEEDYhgSyEEDYhgSyEEDYhgSyEEDYhgSyEEDYhgSyEEDYhgSyEEDYhgSyEEDYhgSyEEDYhgSyEEDYhgSyEEDYhgSyEEDYhgSyEEDYhgSyEEDbx/7dx2ZTyyUbHAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from Bio.ExPASy import cellosaurus \n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "female= 0\n",
    "male = 0\n",
    "unspecified = 0\n",
    "\n",
    "with open(\"../source/cellosaurus.txt\") as handle:\n",
    "    records = cellosaurus.parse(handle)\n",
    "    for record in records:\n",
    "        # Access only to homo sapiens data\n",
    "        if \"Homo sapiens\" in record[\"OX\"][0]:\n",
    "            # Count sex cate\n",
    "            if record[\"SX\"] == \"Female\":\n",
    "                female += 1\n",
    "            elif record[\"SX\"] == \"Male\":\n",
    "                male += 1\n",
    "            elif record[\"SX\"] == \"Sex unspecified\":\n",
    "                unspecified += 1\n",
    "            else:\n",
    "                pass\n",
    "            \n",
    "print(\"Sex ratios for Homo sapiens cells on cellasaurus:\")\n",
    "\n",
    "# Pie chart, where the slices will be ordered and plotted counter-clockwise\n",
    "labels = 'Female', 'Male', 'Unspecified'\n",
    "sizes = [female, male, unspecified]\n",
    "\n",
    "fig1, ax1 = plt.subplots()\n",
    "ax1.pie(sizes, labels=labels, autopct='%1.1f%%', colors = ['lightcoral', 'lightskyblue', 'gold'], shadow=True, startangle=90)\n",
    "ax1.axis('equal')  # Equal aspect ratio ensures that pie is drawn as a circle\n",
    "\n",
    "plt.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
